diff --git a/src/stm32/Kconfig b/src/stm32/Kconfig
index 037e37bb..c6650df2 100644
--- a/src/stm32/Kconfig
+++ b/src/stm32/Kconfig
@@ -111,6 +111,14 @@ choice
         bool "Nation N32G455"
         select MACH_N32G45x
         select MACH_STM32F1
+    config MACH_AT32F415
+        bool "AT32F415"
+        select MACH_STM32F1
+        select MACH_AT32F4x
+    config MACH_AT32F403A
+        bool "AT32F403A"
+        select MACH_STM32F1
+        select MACH_AT32F4x
 endchoice
 
 config MACH_STM32F103x6
@@ -145,13 +153,15 @@ config MACH_STM32L4
     bool
 config MACH_N32G45x
     bool
+config MACH_AT32F4x
+    bool
 config HAVE_STM32_USBFS
     bool
     default y if MACH_STM32F0x2 || MACH_STM32G0Bx || MACH_STM32L4 || MACH_STM32G4
-    default y if (MACH_STM32F1 || MACH_STM32F070) && !STM32_CLOCK_REF_INTERNAL
+    default y if (MACH_STM32F1 || MACH_STM32F070) && !STM32_CLOCK_REF_INTERNAL && !MACH_AT32F415
 config HAVE_STM32_USBOTG
     bool
-    default y if MACH_STM32F2 || MACH_STM32F4 || MACH_STM32F7 || MACH_STM32H7
+    default y if MACH_STM32F2 || MACH_STM32F4 || MACH_STM32F7 || MACH_STM32H7 || MACH_AT32F415
 config HAVE_STM32_CANBUS
     bool
     default y if MACH_STM32F1 || MACH_STM32F2 || MACH_STM32F4x5 || MACH_STM32F446 || MACH_STM32F0x2
@@ -162,7 +172,7 @@ config HAVE_STM32_USBCANBUS
     bool
     depends on HAVE_STM32_USBFS || HAVE_STM32_USBOTG
     depends on HAVE_STM32_CANBUS || HAVE_STM32_FDCANBUS
-    depends on !MACH_STM32F1
+    depends on !MACH_STM32F1 || MACH_AT32F403A
     default y
 
 config MCU
@@ -190,6 +200,8 @@ config MCU
     default "stm32h750xx" if MACH_STM32H750
     default "stm32l412xx" if MACH_STM32L412
     default "stm32f103xe" if MACH_N32G45x
+    default "stm32f105xc" if MACH_AT32F415
+    default "stm32f103xe" if MACH_AT32F403A
 
 config CLOCK_FREQ
     int
@@ -208,6 +220,8 @@ config CLOCK_FREQ
     default 80000000 if MACH_STM32L412
     default 64000000 if MACH_N32G45x && STM32_CLOCK_REF_INTERNAL
     default 128000000 if MACH_N32G45x
+    default 144000000 if MACH_AT32F415
+    default 240000000 if MACH_AT32F403A
 
 config FLASH_SIZE
     hex
@@ -221,6 +235,8 @@ config FLASH_SIZE
     default 0x20000 if MACH_STM32H750
     default 0x200000 if MACH_STM32H743 || MACH_STM32F765
     default 0x20000 if MACH_N32G45x
+    default 0x40000 if MACH_AT32F415
+    default 0x40000 if MACH_AT32F403A
 
 config FLASH_BOOT_ADDRESS
     hex
@@ -248,14 +264,21 @@ config RAM_SIZE
     default 0x24000 if MACH_STM32G0Bx
     default 0x20000 if MACH_STM32H7
     default 0x10000 if MACH_N32G45x
+    default 0x8000 if MACH_AT32F415
+    default 0x18000 if MACH_AT32F403A
 
 config STACK_SIZE
     int
     default 512
 
+config AT32_ENABLE_DEBUG_USART
+    bool "Enable debug serial (Both AT32F415 and AT32F403A use PB10)"
+    depends on (MACH_AT32F415 || MACH_AT32F403A) && LOW_LEVEL_OPTIONS
+    default n
+
 config STM32F103GD_DISABLE_SWD
     bool "Disable SWD at startup (for GigaDevice stm32f103 clones)"
-    depends on MACH_STM32F103 && LOW_LEVEL_OPTIONS
+    depends on (MACH_STM32F103 || MACH_AT32F403A) && LOW_LEVEL_OPTIONS
     default n
     help
         The GigaDevice clone of the STM32F103 may not be able to
@@ -424,6 +447,10 @@ choice
         bool "Serial (on USART5 PD2/PD3)" if LOW_LEVEL_OPTIONS
         depends on MACH_STM32G0Bx
         select SERIAL
+    config STM32_SERIAL_AT_USART5_PB8_PB9
+        bool "Serial (on arterytek_403A USART5 PB8/PB9)" if LOW_LEVEL_OPTIONS
+        depends on MACH_AT32F403A
+        select SERIAL
     config STM32_CANBUS_PA11_PA12
         bool "CAN bus (on PA11/PA12)"
         depends on HAVE_STM32_CANBUS || HAVE_STM32_FDCANBUS
@@ -472,6 +499,11 @@ choice
         bool "USB to CAN bus bridge (USB on PA11/PA12)"
         depends on HAVE_STM32_USBCANBUS
         select USBCANBUS
+    config STM32_USBCANBUS_PA11_PA12_AND_SERIAL_USART5_PB8_PB9
+        bool "USB to CAN bus bridge (USB on PA11/PA12) and use Serial(PB8/PB9), The CAN bus bridge mcu can only be configured as serial mode"
+        depends on (HAVE_STM32_USBCANBUS && MACH_AT32F403A)
+        select USBCANBUS
+        select SERIAL
 endchoice
 choice
     prompt "CAN bus interface" if USBCANBUS
@@ -482,10 +514,10 @@ choice
         depends on HAVE_STM32_CANBUS && MACH_STM32F4
     config STM32_CMENU_CANBUS_PB5_PB6
         bool "CAN bus (on PB5/PB6)"
-        depends on (HAVE_STM32_CANBUS && MACH_STM32F4) || (HAVE_STM32_FDCANBUS && MACH_STM32G0B1)
+        depends on (HAVE_STM32_CANBUS && (MACH_STM32F4 || MACH_AT32F403A)) || (HAVE_STM32_FDCANBUS && MACH_STM32G0B1)
     config STM32_CMENU_CANBUS_PB12_PB13
         bool "CAN bus (on PB12/PB13)"
-        depends on (HAVE_STM32_CANBUS && MACH_STM32F4) || HAVE_STM32_FDCANBUS
+        depends on (HAVE_STM32_CANBUS && (MACH_STM32F4 || MACH_AT32F403A)) || HAVE_STM32_FDCANBUS
     config STM32_CMENU_CANBUS_PD0_PD1
         bool "CAN bus (on PD0/PD1)"
         depends on HAVE_STM32_CANBUS || HAVE_STM32_FDCANBUS
@@ -512,7 +544,7 @@ config STM32_CANBUS_PB5_PB6
     default y if STM32_MMENU_CANBUS_PB5_PB6 || STM32_CMENU_CANBUS_PB5_PB6
 config STM32_CANBUS_PB12_PB13
     bool
-    default y if STM32_MMENU_CANBUS_PB12_PB13 || STM32_CMENU_CANBUS_PB12_PB13
+    default y if STM32_MMENU_CANBUS_PB12_PB13 || STM32_CMENU_CANBUS_PB12_PB13 || STM32_USBCANBUS_PA11_PA12_AND_SERIAL
 config STM32_CANBUS_PD0_PD1
     bool
     default y if STM32_MMENU_CANBUS_PD0_PD1 || STM32_CMENU_CANBUS_PD0_PD1
diff --git a/src/stm32/Makefile b/src/stm32/Makefile
index 5f4d3af5..475ab2da 100644
--- a/src/stm32/Makefile
+++ b/src/stm32/Makefile
@@ -14,6 +14,8 @@ dirs-$(CONFIG_MACH_STM32G0) += lib/stm32g0
 dirs-$(CONFIG_MACH_STM32G4) += lib/stm32g4
 dirs-$(CONFIG_MACH_STM32H7) += lib/stm32h7
 dirs-$(CONFIG_MACH_STM32L4) += lib/stm32l4
+dirs-$(CONFIG_MACH_AT32F415) += lib/at32f415 lib/at32f415/hal/src lib/stm32f1 lib/at32f415/usbd_class/cdc lib/at32f415/usb_drivers/src
+dirs-$(CONFIG_MACH_AT32F403A) += lib/at32f403a lib/at32f403a/hal/src lib/at32f403a/usbd_class/cdc lib/at32f403a/usbd_drivers/src
 
 MCU := $(shell echo $(CONFIG_MCU))
 MCU_UPPER := $(shell echo $(CONFIG_MCU) | tr a-z A-Z | tr X x)
@@ -21,6 +23,10 @@ MCU_UPPER := $(shell echo $(CONFIG_MCU) | tr a-z A-Z | tr X x)
 CFLAGS-$(CONFIG_MACH_STM32F0) += -mcpu=cortex-m0 -Ilib/stm32f0/include
 CFLAGS-$(CONFIG_MACH_STM32F103) += -mcpu=cortex-m3
 CFLAGS-$(CONFIG_MACH_N32G45x) += -mcpu=cortex-m4 -Ilib/n32g45x/include
+CFLAGS-$(CONFIG_MACH_AT32F403A) += -mcpu=cortex-m4 -DAT32F407RCT7 -Ilib/at32f403a -Ilib/at32f403a/hal/inc
+CFLAGS-$(CONFIG_MACH_AT32F403A) += -Ilib/at32f403a/usbd_drivers/inc -Ilib/at32f403a/usbd_class/cdc
+CFLAGS-$(CONFIG_MACH_AT32F415) += -mcpu=cortex-m4 -DAT32F415RCT7 -Ilib/at32f415 -Ilib/at32f415/hal/inc
+CFLAGS-$(CONFIG_MACH_AT32F415) += -Ilib/at32f415/usb_drivers/inc -Ilib/at32f415/usbd_class/cdc
 CFLAGS-$(CONFIG_MACH_STM32F1) += -Ilib/stm32f1/include
 CFLAGS-$(CONFIG_MACH_STM32F2) += -mcpu=cortex-m3 -Ilib/stm32f2/include
 CFLAGS-$(CONFIG_MACH_STM32F4) += -mcpu=cortex-m4 -Ilib/stm32f4/include
@@ -43,6 +49,19 @@ src-$(CONFIG_MACH_STM32F0) += ../lib/stm32f0/system_stm32f0xx.c
 src-$(CONFIG_MACH_STM32F0) += generic/timer_irq.c stm32/stm32f0_timer.c
 src-$(CONFIG_MACH_STM32F0) += stm32/stm32f0.c stm32/gpioperiph.c
 src-$(CONFIG_MACH_STM32F0) += stm32/stm32f0_adc.c stm32/stm32f0_i2c.c
+src-$(CONFIG_MACH_AT32F415) += stm32/at32f415rc.c stm32/adc.c
+src-$(CONFIG_MACH_AT32F415) += ../lib/at32f415/at32f415_clock.c ../lib/at32f415/system_at32f415.c
+src-$(CONFIG_MACH_AT32F415) += ../lib/at32f415/hal/src/at32f415_usart.c ../lib/at32f415/hal/src/at32f415_gpio.c
+src-$(CONFIG_MACH_AT32F415) += ../lib/at32f415/hal/src/at32f415_crm.c ../lib/stm32f1/system_stm32f1xx.c
+src-$(CONFIG_MACH_AT32F415) += ../lib/at32f415/hal/src/at32f415_tmr.c ../lib/at32f415/hal/src/at32f415_flash.c
+src-$(CONFIG_MACH_AT32F403A) += ../lib/stm32f1/system_stm32f1xx.c ../lib/at32f403a/at32f403a_407_clock.c
+src-$(CONFIG_MACH_AT32F403A) += stm32/adc.c stm32/at32f403a.c
+src-$(CONFIG_MACH_AT32F403A) += ../lib/at32f403a/hal/src/at32f403a_407_crm.c ../lib/at32f403a/hal/src/at32f403a_407_tmr.c
+src-$(CONFIG_MACH_AT32F403A) += ../lib/at32f403a/hal/src/at32f403a_407_gpio.c ../lib/at32f403a/hal/src/at32f403a_407_flash.c
+src-$(CONFIG_MACH_AT32F403A) += ../lib/at32f403a/hal/src/at32f403a_407_usart.c
+src-$(CONFIG_MACH_AT32F403A) += ../lib/at32f403a/hal/src/at32f403a_407_acc.c
+src-$(CONFIG_MACH_AT32F403A) += ../lib/at32f403a/system_at32f403a_407.c
+src-$(CONFIG_MACH_AT32F4x) += stm32/inductance_coil.c stm32/power_loss_check.c
 src-$(CONFIG_MACH_STM32F103) += ../lib/stm32f1/system_stm32f1xx.c
 src-$(CONFIG_MACH_STM32F103) += stm32/adc.c
 src-$(CONFIG_MACH_N32G45x) += ../lib/stm32f1/system_stm32f1xx.c
diff --git a/src/stm32/at32f403a.c b/src/stm32/at32f403a.c
new file mode 100644
index 00000000..3a6dc2bb
--- /dev/null
+++ b/src/stm32/at32f403a.c
@@ -0,0 +1,184 @@
+#include "at32f403a.h"
+
+#include "at32f403a_407.h"
+#include "at32f403a_407_clock.h"
+
+#include "sched.h" // DECL_INIT
+#include "command.h" // DECL_CONSTANT_STR
+#include "board/armcm_boot.h" // armcm_enable_irq
+#include "board/armcm_timer.h" // udelay
+
+#include <string.h> // memmove
+
+/**************** define print uart ******************/
+#define PRINT_UART                       USART3
+#define PRINT_UART_CRM_CLK               CRM_USART3_PERIPH_CLOCK
+#define PRINT_UART_TX_PIN                GPIO_PINS_10
+#define PRINT_UART_TX_GPIO               GPIOB
+#define PRINT_UART_TX_GPIO_CRM_CLK       CRM_GPIOB_PERIPH_CLOCK
+// #define PRINT_UART_GPIO_REMAP            USART6_GMUX
+
+#if CONFIG_AT32_ENABLE_DEBUG_USART
+  DECL_CONSTANT_STR("RESERVE_PINS_debug_uart_tx_pin", "PB10");
+#endif
+
+#if defined (__GNUC__) && !defined (__clang__)
+  #define PUTCHAR_PROTOTYPE int __io_putchar(int ch)
+#else
+  #define PUTCHAR_PROTOTYPE int fputc(int ch, FILE *f)
+#endif
+
+
+/**
+  * @brief  retargets the c library printf function to the usart.
+  * @param  none
+  * @retval none
+  */
+PUTCHAR_PROTOTYPE
+{
+  while(usart_flag_get(PRINT_UART, USART_TDBE_FLAG) == RESET);
+  usart_data_transmit(PRINT_UART, (uint16_t)ch);
+  while(usart_flag_get(PRINT_UART, USART_TDC_FLAG) == RESET);
+  return ch;
+}
+
+void at32f403a_log(char* log)
+{
+  while (log && *log) {
+    while(usart_flag_get(PRINT_UART, USART_TDBE_FLAG) == RESET);
+    usart_data_transmit(PRINT_UART, (uint16_t)(*log));
+    log++;
+  }
+}
+
+
+/**
+  * @brief  initialize uart
+  * @param  baudrate: uart baudrate
+  * @retval none
+  */
+void uart_debug_print_init(uint32_t baudrate)
+{
+  gpio_init_type gpio_init_struct;
+
+// #if defined (__GNUC__) && !defined (__clang__)
+//   setvbuf(stdout, NULL, _IONBF, 0);
+// #endif
+
+  /* enable the uart and gpio clock */
+  crm_periph_clock_enable(PRINT_UART_CRM_CLK, TRUE);
+  crm_periph_clock_enable(PRINT_UART_TX_GPIO_CRM_CLK, TRUE);
+
+#ifdef PRINT_UART_GPIO_REMAP
+  gpio_pin_remap_config(PRINT_UART_GPIO_REMAP, TRUE);
+#endif
+
+  gpio_default_para_init(&gpio_init_struct);
+
+  /* configure the uart tx pin */
+  gpio_init_struct.gpio_drive_strength = GPIO_DRIVE_STRENGTH_STRONGER;
+  gpio_init_struct.gpio_out_type  = GPIO_OUTPUT_PUSH_PULL;
+  gpio_init_struct.gpio_mode = GPIO_MODE_MUX;
+  gpio_init_struct.gpio_pins = PRINT_UART_TX_PIN;
+  gpio_init_struct.gpio_pull = GPIO_PULL_NONE;
+  gpio_init(PRINT_UART_TX_GPIO, &gpio_init_struct);
+
+  /* configure uart param */
+  usart_init(PRINT_UART, baudrate, USART_DATA_8BITS, USART_STOP_1_BIT);
+  usart_transmitter_enable(PRINT_UART, TRUE);
+  usart_enable(PRINT_UART, TRUE);
+}
+
+
+/**
+  * @brief  usb 48M clock select
+  * @param  clk_s:USB_CLK_HICK, USB_CLK_HEXT
+  * @retval none
+  */
+static void usb_clock48m_select(usb_clk48_s clk_s)
+{
+  if(clk_s == USB_CLK_HICK)
+  {
+    crm_usb_clock_source_select(CRM_USB_CLOCK_SOURCE_HICK);
+
+    /* enable the acc calibration ready interrupt */
+    crm_periph_clock_enable(CRM_ACC_PERIPH_CLOCK, TRUE);
+
+    /* update the c1\c2\c3 value */
+    acc_write_c1(7980);
+    acc_write_c2(8000);
+    acc_write_c3(8020);
+
+    /* open acc calibration */
+    acc_calibration_mode_enable(ACC_CAL_HICKTRIM, TRUE);
+  }
+  else
+  {
+    switch(SystemCoreClock)
+    {
+      /* 48MHz */
+      case 48000000:
+        crm_usb_clock_div_set(CRM_USB_DIV_1);
+        break;
+
+      /* 72MHz */
+      case 72000000:
+        crm_usb_clock_div_set(CRM_USB_DIV_1_5);
+        break;
+
+      /* 96MHz */
+      case 96000000:
+        crm_usb_clock_div_set(CRM_USB_DIV_2);
+        break;
+
+      /* 120MHz */
+      case 120000000:
+        crm_usb_clock_div_set(CRM_USB_DIV_2_5);
+        break;
+
+      /* 144MHz */
+      case 144000000:
+        crm_usb_clock_div_set(CRM_USB_DIV_3);
+        break;
+
+      /* 168MHz */
+      case 168000000:
+        crm_usb_clock_div_set(CRM_USB_DIV_3_5);
+        break;
+
+      /* 192MHz */
+      case 192000000:
+        crm_usb_clock_div_set(CRM_USB_DIV_4);
+        break;
+
+      default:
+        break;
+
+    }
+  }
+}
+
+#if CONFIG_STM32_SERIAL_AT_USART5_PB8_PB9 || CONFIG_STM32_USBCANBUS_PA11_PA12_AND_SERIAL_USART5_PB8_PB9
+void mcu_uart_gpio_remap(void) {
+  gpio_pin_remap_config(UART5_GMUX_0001, TRUE);
+}
+#endif
+
+void mcu_can2_gpio_remap(void) {
+  gpio_pin_remap_config(CAN2_GMUX_0001, TRUE);
+}
+
+void mcu_spi4_gpio_remap(void) {
+  gpio_pin_remap_config(SPI4_GMUX_0001, TRUE);
+}
+
+void at32f403a_clock_setup(void)
+{
+  system_clock_config();
+
+  /* select usb 48m clcok source */
+  usb_clock48m_select(USB_CLK_HICK);
+
+  /* enable usb clock */
+  crm_periph_clock_enable(CRM_USB_PERIPH_CLOCK, TRUE);
+}
diff --git a/src/stm32/at32f403a.h b/src/stm32/at32f403a.h
new file mode 100644
index 00000000..48b57cea
--- /dev/null
+++ b/src/stm32/at32f403a.h
@@ -0,0 +1,38 @@
+#ifndef __STM32_AT32F403A_H
+#define __STM32_AT32F403A_H
+
+#include "autoconf.h"
+#include <stdint.h>
+
+void at32f403a_clock_setup(void);
+void uart_debug_print_init(uint32_t baudrate);
+void mcu_can2_gpio_remap(void);
+void mcu_spi4_gpio_remap(void);
+#if CONFIG_STM32_SERIAL_AT_USART5_PB8_PB9 || CONFIG_STM32_USBCANBUS_PA11_PA12_AND_SERIAL_USART5_PB8_PB9
+void mcu_uart_gpio_remap(void);
+#endif
+
+#ifndef TAG
+  #define TAG
+#endif
+
+#if CONFIG_MACH_AT32F403A
+  void at32f403a_log(char* log);
+  // #define LOG_E(format, ...) at32f403a_log("[E]" TAG format, ##__VA_ARGS__)
+  // #define LOG_W(format, ...) at32f403a_log("[W]" TAG format, ##__VA_ARGS__)
+  // #define LOG_I(format, ...) at32f403a_log("[I]" TAG format, ##__VA_ARGS__)
+  // #define LOG_V(format, ...) at32f403a_log("[V]" TAG format, ##__VA_ARGS__)
+
+  #define LOG_E(format) at32f403a_log("[E]" TAG format)
+  #define LOG_W(format) at32f403a_log("[W]" TAG format)
+  #define LOG_I(format) at32f403a_log("[I]" TAG format)
+  #define LOG_V(format) at32f403a_log("[V]" TAG format)
+#else
+  #define LOG_E(format, ...)
+  #define LOG_W(format, ...)
+  #define LOG_I(format, ...)
+  #define LOG_V(format, ...)
+
+#endif
+
+#endif
diff --git a/src/stm32/at32f415rc.c b/src/stm32/at32f415rc.c
new file mode 100644
index 00000000..ee3764f8
--- /dev/null
+++ b/src/stm32/at32f415rc.c
@@ -0,0 +1,118 @@
+#include "at32f415.h"
+#include "at32f415rc.h"
+#include "at32f415_clock.h"
+#include "command.h" // DECL_CONSTANT_STR
+
+/**************** define print uart ******************/
+#define PRINT_UART                       USART3
+#define PRINT_UART_CRM_CLK               CRM_USART3_PERIPH_CLOCK
+#define PRINT_UART_TX_PIN                GPIO_PINS_10
+#define PRINT_UART_TX_GPIO               GPIOB
+#define PRINT_UART_TX_GPIO_CRM_CLK       CRM_GPIOB_PERIPH_CLOCK
+
+#if CONFIG_AT32_ENABLE_DEBUG_USART
+  DECL_CONSTANT_STR("RESERVE_PINS_debug_uart_tx_pin", "PB10");
+#endif
+
+void at32f415_log(char* log)
+{
+  while (log && *log) {
+    while(usart_flag_get(PRINT_UART, USART_TDBE_FLAG) == RESET);
+    usart_data_transmit(PRINT_UART, (uint16_t)(*log));
+    log++;
+  }
+}
+
+/**
+  * @brief  initialize uart
+  * @param  baudrate: uart baudrate
+  * @retval none
+  */
+void uart_debug_print_init(uint32_t baudrate)
+{
+  gpio_init_type gpio_init_struct;
+
+  /* enable the uart and gpio clock */
+  crm_periph_clock_enable(PRINT_UART_CRM_CLK, TRUE);
+  crm_periph_clock_enable(PRINT_UART_TX_GPIO_CRM_CLK, TRUE);
+
+  gpio_default_para_init(&gpio_init_struct);
+
+  /* configure the uart tx pin */
+  gpio_init_struct.gpio_drive_strength = GPIO_DRIVE_STRENGTH_STRONGER;
+  gpio_init_struct.gpio_out_type  = GPIO_OUTPUT_PUSH_PULL;
+  gpio_init_struct.gpio_mode = GPIO_MODE_MUX;
+  gpio_init_struct.gpio_pins = PRINT_UART_TX_PIN;
+  gpio_init_struct.gpio_pull = GPIO_PULL_NONE;
+  gpio_init(PRINT_UART_TX_GPIO, &gpio_init_struct);
+
+  /* configure uart param */
+  usart_init(PRINT_UART, baudrate, USART_DATA_8BITS, USART_STOP_1_BIT);
+  usart_transmitter_enable(PRINT_UART, TRUE);
+  usart_enable(PRINT_UART, TRUE);
+}
+
+
+
+/****************************************************************
+ * USB init
+ ****************************************************************/
+#include "usb_conf.h"
+
+/**
+  * @brief  usb 48M clock select
+  * @param  clk_s:USB_CLK_HICK, USB_CLK_HEXT
+  * @retval none
+  */
+void usb_clock48m_select(usb_clk48_s clk_s)
+{
+  switch(system_core_clock)
+  {
+    /* 48MHz */
+    case 48000000:
+      crm_usb_clock_div_set(CRM_USB_DIV_1);
+      break;
+
+    /* 72MHz */
+    case 72000000:
+      crm_usb_clock_div_set(CRM_USB_DIV_1_5);
+      break;
+
+    /* 96MHz */
+    case 96000000:
+      crm_usb_clock_div_set(CRM_USB_DIV_2);
+      break;
+
+    /* 120MHz */
+    case 120000000:
+      crm_usb_clock_div_set(CRM_USB_DIV_2_5);
+      break;
+
+    /* 144MHz */
+    case 144000000:
+      crm_usb_clock_div_set(CRM_USB_DIV_3);
+      break;
+
+    default:
+      break;
+  }
+}
+
+/****************************************************************
+ * customizing External Reference Functions
+ ****************************************************************/
+
+void at32f415rc_clock_setup(void)
+{
+  system_clock_config();
+}
+
+void at32f415rc_usbotg_clock_config(void)
+{
+  /* enable otgfs clock */
+  crm_periph_clock_enable(OTG_CLOCK, TRUE);
+
+  /* select usb 48m clcok source */
+  usb_clock48m_select(USB_CLK_HEXT);
+
+}
diff --git a/src/stm32/at32f415rc.h b/src/stm32/at32f415rc.h
new file mode 100644
index 00000000..955bcc4f
--- /dev/null
+++ b/src/stm32/at32f415rc.h
@@ -0,0 +1,27 @@
+#ifndef __STM32_AT32F415RC_H
+#define __STM32_AT32F415RC_H
+
+void at32f415rc_clock_setup(void);
+void uart_debug_print_init(uint32_t baudrate);
+void at32f415rc_usbotg_clock_config(void);
+
+#ifndef TAG
+  #define TAG
+#endif
+
+#if CONFIG_MACH_AT32F415
+  void at32f415_log(char* log);
+
+  #define LOG_E(format) at32f415_log("[E] " TAG format)
+  #define LOG_W(format) at32f415_log("[W] " TAG format)
+  #define LOG_I(format) at32f415_log("[I] " TAG format)
+  #define LOG_V(format) at32f415_log("[V] " TAG format)
+#else
+  #define LOG_E(format, ...)
+  #define LOG_W(format, ...)
+  #define LOG_I(format, ...)
+  #define LOG_V(format, ...)
+
+#endif
+
+#endif
\ No newline at end of file
diff --git a/src/stm32/can.c b/src/stm32/can.c
index dc915328..3169194d 100644
--- a/src/stm32/can.c
+++ b/src/stm32/can.c
@@ -58,12 +58,34 @@
 #endif
 
 #if CONFIG_MACH_STM32F1
- #define SOC_CAN CAN1
- #define FILTER_CAN CAN1
- #define CAN_RX0_IRQn  CAN1_RX0_IRQn
- #define CAN_RX1_IRQn  CAN1_RX1_IRQn
- #define CAN_TX_IRQn   CAN1_TX_IRQn
- #define CAN_SCE_IRQn  CAN1_SCE_IRQn
+ #if CONFIG_MACH_AT32F403A
+  #if (CONFIG_STM32_CANBUS_PA11_PA12 || CONFIG_STM32_CANBUS_PB8_PB9 \
+    || CONFIG_STM32_CANBUS_PD0_PD1 || CONFIG_STM32_CANBUS_PI9_PH13)
+   #define SOC_CAN CAN1
+   #define FILTER_CAN CAN1
+   #define CAN_RX0_IRQn  CAN1_RX0_IRQn
+   #define CAN_RX1_IRQn  CAN1_RX1_IRQn
+   #define CAN_TX_IRQn   CAN1_TX_IRQn
+   #define CAN_SCE_IRQn  CAN1_SE_IRQn
+  #elif CONFIG_STM32_CANBUS_PB5_PB6 || CONFIG_STM32_CANBUS_PB12_PB13
+   #define CAN2          ((CAN_TypeDef *)(APB1PERIPH_BASE + 0x6800))
+   #define SOC_CAN       CAN2
+   #define FILTER_CAN    CAN2
+   #define CAN_RX0_IRQn  69 //CAN2_RX0_IRQn
+   #define CAN_RX1_IRQn  70 //CAN2_RX1_IRQn
+   #define CAN_TX_IRQn   68 //CAN2_TX_IRQn
+   #define CAN_SCE_IRQn  71 //CAN2_SE_IRQn
+  #else
+   #error Uknown pins for AT32F403A CAN
+  #endif
+ #else
+  #define SOC_CAN CAN1
+  #define FILTER_CAN CAN1
+  #define CAN_RX0_IRQn  CAN1_RX0_IRQn
+  #define CAN_RX1_IRQn  CAN1_RX1_IRQn
+  #define CAN_TX_IRQn   CAN1_TX_IRQn
+  #define CAN_SCE_IRQn  CAN1_SCE_IRQn
+ #endif
  #define CAN_FUNCTION  GPIO_FUNCTION(9)  // Alternative function mapping number
 #endif
 
@@ -256,6 +278,11 @@ can_init(void)
         // Also enable CAN1 clock if using CAN2 with filter on CAN1
         enable_pclock((uint32_t)FILTER_CAN);
 
+    #if CONFIG_MACH_AT32F403A && CONFIG_STM32_CANBUS_PB5_PB6
+        extern void mcu_can2_gpio_remap(void);
+        mcu_can2_gpio_remap();
+    #endif
+
     gpio_peripheral(GPIO_Rx, CAN_FUNCTION, 1);
     gpio_peripheral(GPIO_Tx, CAN_FUNCTION, 0);
 
diff --git a/src/stm32/inductance_coil.c b/src/stm32/inductance_coil.c
new file mode 100644
index 00000000..a63a820e
--- /dev/null
+++ b/src/stm32/inductance_coil.c
@@ -0,0 +1,674 @@
+#include <string.h> // ffs
+#include "board/irq.h" // irq_save
+#include "board/misc.h"
+#include "board/armcm_boot.h" // armcm_enable_irq
+#include "command.h" // DECL_ENUMERATION_RANGE
+#include "basecmd.h" // oid_alloc
+#include "sched.h" // sched_shutdown
+#include "autoconf.h" // CONFIG_CLOCK_FREQ
+#include "inductance_coil.h"
+#include "../sensor_bulk.h"
+
+#define BYTES_PER_SAMPLE (4)
+#define MAX_BUFFER_SIZE 50
+
+typedef enum {
+  OPEN = 0,
+  TRIGGERED = 1,
+} VIRTUAL_GPIO_STATE;
+
+typedef enum {
+  FIXED_TIME_CAL_MODE = 0,
+  FIXED_PULSE_NUM_CAL_MODE = 1,
+} FREQ_CAL_MODE;
+
+typedef struct {
+  uint32_t buffer[MAX_BUFFER_SIZE];
+  uint32_t bufferSize;
+  uint32_t head;
+  uint32_t count;
+  uint32_t sum;
+} MOVING_SUM;
+
+struct freq_cal_info {
+  uint8_t trigger_mode;     // 0: Unidirectional Trigger  1:  Bidirectional Trigger
+  uint8_t trigger_invert;
+  uint32_t need_capture_pulse_sum;
+  uint32_t cal_capture_pulse_window_size;
+  uint32_t trg_freq_ht;
+  uint32_t trg_freq_lt;
+  volatile uint32_t cmp_trg_cnt_ht;
+  volatile uint32_t cmp_trg_cnt_lt;
+  volatile uint32_t capture_pulse_sum;
+  volatile uint32_t update_cnt;
+  double freq_cal_cycle;
+  double freq_cal_timeout_cycle;
+  double freq_cal_factor;
+  volatile VIRTUAL_GPIO_STATE virtual_gpio_state;
+  FREQ_CAL_MODE freq_cal_mode;
+  MOVING_SUM moving_sum;
+};
+
+struct timer_config_freq_param {
+  uint8_t oid;
+  uint8_t absolute_mode;
+  uint8_t trigger_mode;
+  uint8_t trigger_invert;
+  uint8_t force_update;
+  int32_t trg_freq_ht;
+  int32_t trg_freq_lt;
+};
+
+struct inductance_coil_dev {
+  struct timer timer;
+  struct timer config_freq_timer;
+  uint16_t dev_flags;
+  uint32_t rest_tick;
+  struct freq_cal_info cal_info;
+  struct sensor_bulk sb;
+  struct timer_config_freq_param config_freq_param;
+};
+
+static uint8_t inductance_coil_oid_malloc = 0;
+static VIRTUAL_GPIO_STATE tmp_gpio_state = TRIGGERED;
+static struct task_wake inductance_coil_wake;
+struct freq_cal_info g_freq_cal_info = {
+                                          .trigger_mode = 0,
+                                          .trigger_invert = 0,
+                                          .trg_freq_ht = 0,
+                                          .trg_freq_lt = 0,
+                                          .cmp_trg_cnt_ht = 0,
+                                          .cmp_trg_cnt_lt = 0,
+                                          .capture_pulse_sum = 0,
+                                          .freq_cal_factor = 1,
+                                          .freq_cal_cycle = 0.001,
+                                          .freq_cal_timeout_cycle = 1,
+                                          .freq_cal_mode = FIXED_TIME_CAL_MODE,
+                                          .need_capture_pulse_sum = 1000,
+                                          .virtual_gpio_state = TRIGGERED,
+                                          .update_cnt = 0,
+                                          .cal_capture_pulse_window_size = 1,
+                                        };
+
+// For faster performance no legality checking is done, the call needs to ensure that the parameters are legal
+void init_moving_sum(MOVING_SUM *ms, uint32_t bufferSize) {
+  if (bufferSize == 0 || bufferSize > MAX_BUFFER_SIZE) {
+    shutdown("Invalid buff size!!!");
+  }
+  ms->bufferSize = bufferSize;
+  ms->head = 0;
+  ms->count = 0;
+  ms->sum = 0;
+}
+
+void add_value(MOVING_SUM *ms, uint32_t value) {
+  if (ms->count >= ms->bufferSize) {
+    ms->sum -= ms->buffer[ms->head];
+  } else {
+    ++ms->count;
+  }
+
+  ms->sum += value;
+  ms->buffer[ms->head] = value;
+
+  if (++ms->head >= ms->bufferSize) {
+    ms->head = 0;
+  }
+}
+
+void reset_buffer(MOVING_SUM *ms) {
+  for (uint32_t i = 0; i < ms->bufferSize; ++i) {
+    ms->buffer[i] = 0;
+  }
+  ms->head = 0;
+  ms->count = 0;
+  ms->sum = 0;
+}
+
+uint32_t get_sum(MOVING_SUM *ms) {
+  if (ms->count + 1 < ms->bufferSize || ms->bufferSize == 0)
+    return 0;
+  return ms->sum;
+}
+
+#if CONFIG_MACH_AT32F4x
+// currently it only works with the AT32F415 chip from artery.
+  #if CONFIG_MACH_AT32F415
+    #include "at32f415_crm.h"
+    #include "at32f415_tmr.h"
+  #elif CONFIG_MACH_AT32F403A
+    #include "at32f403a_407_crm.h"
+    #include "at32f403a_407_tmr.h"
+  #endif
+
+// use channel 1 of timer2, expand for more channels and timers in the future
+// DECL_CONSTANT_STR("RESERVE_PINS_input_capture", "PA0");
+#define INPUT_CAPTURE_CRM_TIM                     TMR2
+#define INPUT_CAPTURE_CRM_TIM_MAX_PR_VALUE        0xFFFFFFFF
+#define INPUT_CAPTURE_CRM_CLK                     CRM_TMR2_PERIPH_CLOCK
+
+#define INPUT_CAPTURE_GPIO_CRM_CLK                CRM_GPIOA_PERIPH_CLOCK
+#define INPUT_CAPTURE_PIN                         GPIO_PINS_0
+#define INPUT_CAPTURE_GPIO                        GPIOA
+#define INVALID_FREQ_VALUE                        0xFFFFFFFF
+#define DAFAULT_CAL_CYCLE_TIME                    (0.001)     // calculate the capture frequency once in 1ms
+#define DAFAULT_TIM_PR_VALUE                      (CONFIG_CLOCK_FREQ * DAFAULT_CAL_CYCLE_TIME)
+
+#define INPUT_CAPTURE_CAL_CRM_TIM                 TMR5
+#define INPUT_CAPTURE_CAL_CRM_TIM_CLOCK_FREQ      CONFIG_CLOCK_FREQ
+#define INPUT_CAPTURE_CAL_CRM_CLK                 CRM_TMR5_PERIPH_CLOCK
+#define INPUT_CAPTURE_CAL_CRM_DIV                 (1)
+#define INPUT_CAPTURE_CAL_CRM_TIM_MAX_PR_VALUE    0xFFFFFFFF
+#define INPUT_CAPTURE_CAL_CRM_TIM_PR(CYCLE)       ((double)CONFIG_CLOCK_FREQ / INPUT_CAPTURE_CAL_CRM_DIV * CYCLE - 1)
+#define INPUT_CAPTURE_OVER_CNT                    (1000)
+#define INPUT_CAPTURE_CAL_IRQn                    (TMR5_GLOBAL_IRQn)
+#define INPUT_CAPTURE_CAL_COMPENSATION_TICK       (0)
+
+void
+InputCapture_IRQHandler(void)
+{
+  if (INPUT_CAPTURE_CAL_CRM_TIM->ists & TMR_OVF_FLAG)
+  {
+    INPUT_CAPTURE_CAL_CRM_TIM->ists = ~TMR_OVF_FLAG;
+    // if (g_freq_cal_info.freq_cal_mode == FIXED_PULSE_NUM_CAL_MODE)
+    //   g_freq_cal_info.capture_pulse_sum = 0;
+    // else
+    //   g_freq_cal_info.capture_pulse_sum = INPUT_CAPTURE_CRM_TIM->cval;
+    if (g_freq_cal_info.freq_cal_mode == FIXED_PULSE_NUM_CAL_MODE) {
+      reset_buffer(&g_freq_cal_info.moving_sum);
+    }
+    else {
+      add_value(&g_freq_cal_info.moving_sum, INPUT_CAPTURE_CRM_TIM->cval);
+    }
+    INPUT_CAPTURE_CRM_TIM->cval =  0;
+  }
+
+  if (INPUT_CAPTURE_CAL_CRM_TIM->ists & TMR_C1_FLAG) {
+    INPUT_CAPTURE_CAL_CRM_TIM->ists = ~TMR_C1_FLAG;
+    if (g_freq_cal_info.freq_cal_mode == FIXED_PULSE_NUM_CAL_MODE)
+      add_value(&g_freq_cal_info.moving_sum, INPUT_CAPTURE_CAL_CRM_TIM->c1dt);
+    // if (g_freq_cal_info.freq_cal_mode == FIXED_PULSE_NUM_CAL_MODE)
+    //   g_freq_cal_info.capture_pulse_sum = INPUT_CAPTURE_CAL_CRM_TIM->c1dt;
+  }
+
+  g_freq_cal_info.capture_pulse_sum = get_sum(&g_freq_cal_info.moving_sum);
+
+  if (g_freq_cal_info.capture_pulse_sum != 0) {
+    if (g_freq_cal_info.freq_cal_mode == FIXED_PULSE_NUM_CAL_MODE) {
+      if (g_freq_cal_info.trigger_mode) {
+        if (g_freq_cal_info.capture_pulse_sum + INPUT_CAPTURE_CAL_COMPENSATION_TICK <= g_freq_cal_info.cmp_trg_cnt_ht || \
+            g_freq_cal_info.capture_pulse_sum + INPUT_CAPTURE_CAL_COMPENSATION_TICK >= g_freq_cal_info.cmp_trg_cnt_lt)
+          tmp_gpio_state = TRIGGERED;
+        else
+          tmp_gpio_state = OPEN;
+      }
+      else {
+        if (g_freq_cal_info.capture_pulse_sum + INPUT_CAPTURE_CAL_COMPENSATION_TICK <= g_freq_cal_info.cmp_trg_cnt_ht)
+          tmp_gpio_state = TRIGGERED;
+        else
+          tmp_gpio_state = OPEN;
+      }
+    }
+    else {
+      if (g_freq_cal_info.trigger_mode) {
+        if (g_freq_cal_info.capture_pulse_sum + INPUT_CAPTURE_CAL_COMPENSATION_TICK >= g_freq_cal_info.cmp_trg_cnt_ht || \
+            g_freq_cal_info.capture_pulse_sum + INPUT_CAPTURE_CAL_COMPENSATION_TICK <= g_freq_cal_info.cmp_trg_cnt_lt)
+          tmp_gpio_state = TRIGGERED;
+        else
+          tmp_gpio_state = OPEN;
+      }
+      else {
+        if (g_freq_cal_info.capture_pulse_sum + INPUT_CAPTURE_CAL_COMPENSATION_TICK >= g_freq_cal_info.cmp_trg_cnt_ht)
+          tmp_gpio_state = TRIGGERED;
+        else
+          tmp_gpio_state = OPEN;
+      }
+    }
+
+    if (g_freq_cal_info.trigger_invert)
+      g_freq_cal_info.virtual_gpio_state = !tmp_gpio_state;
+    else
+      g_freq_cal_info.virtual_gpio_state = tmp_gpio_state;
+  }
+  else {
+    g_freq_cal_info.virtual_gpio_state = TRIGGERED;
+  }
+  g_freq_cal_info.update_cnt++;
+}
+
+void
+crm_configuration(void)
+{
+  // input capture tmr clock enable
+  crm_periph_clock_enable(INPUT_CAPTURE_CRM_CLK, TRUE);
+
+  // freq cal tmr clock enable
+  crm_periph_clock_enable(INPUT_CAPTURE_CAL_CRM_CLK, TRUE);
+
+  // gpio clock enable
+  crm_periph_clock_enable(INPUT_CAPTURE_GPIO_CRM_CLK, TRUE);
+}
+
+void
+gpio_configuration(void)
+{
+  gpio_init_type  gpio_init_struct = {0};
+  gpio_init_struct.gpio_pins = INPUT_CAPTURE_PIN;
+  gpio_init_struct.gpio_mode = GPIO_MODE_INPUT;
+  gpio_init_struct.gpio_out_type = GPIO_OUTPUT_PUSH_PULL;
+  gpio_init_struct.gpio_pull = GPIO_PULL_DOWN;
+  gpio_init_struct.gpio_drive_strength = GPIO_DRIVE_STRENGTH_STRONGER;
+  gpio_init(INPUT_CAPTURE_GPIO, &gpio_init_struct);
+}
+
+void
+inductance_coil_crm_tmr_init(struct freq_cal_info *info)
+{
+  uint32_t tmp_pr_value = 0;
+
+  // turn off the timer.
+  tmr_counter_enable(INPUT_CAPTURE_CRM_TIM, FALSE);
+  tmr_counter_enable(INPUT_CAPTURE_CAL_CRM_TIM, FALSE);
+  tmr_interrupt_enable(INPUT_CAPTURE_CAL_CRM_TIM, TMR_OVF_INT | TMR_C1_INT, FALSE);
+
+  // reset the flags
+  INPUT_CAPTURE_CRM_TIM->ists &= 0;
+  INPUT_CAPTURE_CAL_CRM_TIM->ists &= 0;
+
+  // cal info init
+  g_freq_cal_info.capture_pulse_sum = 0;
+  g_freq_cal_info.need_capture_pulse_sum = info->need_capture_pulse_sum;
+  g_freq_cal_info.freq_cal_cycle = info->freq_cal_cycle;
+  g_freq_cal_info.freq_cal_timeout_cycle = info->freq_cal_timeout_cycle;
+  g_freq_cal_info.virtual_gpio_state = TRIGGERED;
+  g_freq_cal_info.update_cnt = 0;
+  g_freq_cal_info.freq_cal_mode = info->freq_cal_mode;
+  // g_freq_cal_info.trigger_freq = info->trigger_freq;
+  g_freq_cal_info.trigger_mode = info->trigger_mode;
+  g_freq_cal_info.trg_freq_ht = info->trg_freq_ht;
+  g_freq_cal_info.trg_freq_lt = info->trg_freq_lt;
+  g_freq_cal_info.trigger_invert = info->trigger_invert;
+  g_freq_cal_info.cal_capture_pulse_window_size = info->cal_capture_pulse_window_size;
+
+  init_moving_sum(&g_freq_cal_info.moving_sum, g_freq_cal_info.cal_capture_pulse_window_size);
+  if (g_freq_cal_info.trigger_mode == 1 && g_freq_cal_info.freq_cal_mode == FIXED_TIME_CAL_MODE)
+    shutdown("fixed time cal mode cannot set bidirectional trigger!!!");
+
+  if (g_freq_cal_info.freq_cal_mode == FIXED_TIME_CAL_MODE) {
+    g_freq_cal_info.cmp_trg_cnt_ht = g_freq_cal_info.trg_freq_ht * g_freq_cal_info.freq_cal_cycle * g_freq_cal_info.cal_capture_pulse_window_size;
+    g_freq_cal_info.cmp_trg_cnt_lt = g_freq_cal_info.trg_freq_lt * g_freq_cal_info.freq_cal_cycle * g_freq_cal_info.cal_capture_pulse_window_size;
+  }
+  else {
+    if (g_freq_cal_info.trigger_mode == 1 && (g_freq_cal_info.trg_freq_ht < g_freq_cal_info.trg_freq_lt))
+      shutdown("trg freq ht is smaller than trg freq lt!!!");
+
+    if (g_freq_cal_info.trg_freq_ht) {
+      g_freq_cal_info.cmp_trg_cnt_ht = ((double)INPUT_CAPTURE_CAL_CRM_TIM_CLOCK_FREQ /
+                                          g_freq_cal_info.trg_freq_ht) * g_freq_cal_info.need_capture_pulse_sum * g_freq_cal_info.cal_capture_pulse_window_size;
+    }
+    else {
+      g_freq_cal_info.cmp_trg_cnt_ht = 0xFFFFFFFF;
+    }
+
+    if (g_freq_cal_info.trg_freq_lt) {
+      g_freq_cal_info.cmp_trg_cnt_lt = ((double)INPUT_CAPTURE_CAL_CRM_TIM_CLOCK_FREQ /
+                                          g_freq_cal_info.trg_freq_lt) * g_freq_cal_info.need_capture_pulse_sum * g_freq_cal_info.cal_capture_pulse_window_size;
+    }
+    else {
+      g_freq_cal_info.cmp_trg_cnt_lt = 0;
+    }
+  }
+
+  // expanded to 32-bit timer
+  tmr_input_config_type  tmr_input_config_struct;
+  tmr_32_bit_function_enable(INPUT_CAPTURE_CRM_TIM, TRUE);
+  tmr_32_bit_function_enable(INPUT_CAPTURE_CAL_CRM_TIM, TRUE);
+
+  if (g_freq_cal_info.freq_cal_mode == FIXED_TIME_CAL_MODE) {
+    tmp_pr_value = INPUT_CAPTURE_CAL_CRM_TIM_PR(g_freq_cal_info.freq_cal_cycle);
+    if (tmp_pr_value > INPUT_CAPTURE_CAL_CRM_TIM_MAX_PR_VALUE) {
+      shutdown("input capture calculation cycle is too long!!!");
+    }
+    tmr_base_init(INPUT_CAPTURE_CAL_CRM_TIM, tmp_pr_value, INPUT_CAPTURE_CAL_CRM_DIV -1);
+    tmr_counter_value_set(INPUT_CAPTURE_CAL_CRM_TIM, 0);
+    tmr_cnt_dir_set(INPUT_CAPTURE_CAL_CRM_TIM, TMR_COUNT_UP);
+     g_freq_cal_info.freq_cal_factor = 1 / (info->freq_cal_cycle * 1 * g_freq_cal_info.cal_capture_pulse_window_size);
+
+    tmr_base_init(INPUT_CAPTURE_CRM_TIM, INPUT_CAPTURE_CRM_TIM_MAX_PR_VALUE, 0);
+    tmr_counter_value_set(INPUT_CAPTURE_CRM_TIM, 0);
+    tmr_cnt_dir_set(INPUT_CAPTURE_CRM_TIM, TMR_COUNT_UP);
+    tmr_external_clock_mode1_config(INPUT_CAPTURE_CRM_TIM, TMR_ES_FREQUENCY_DIV_1, TMR_ES_POLARITY_NON_INVERTED, 0);
+  }
+  else {
+    tmr_primary_mode_select(INPUT_CAPTURE_CRM_TIM, TMR_PRIMARY_SEL_OVERFLOW);
+    tmr_sub_sync_mode_set(INPUT_CAPTURE_CRM_TIM, TRUE);
+    tmr_base_init(INPUT_CAPTURE_CRM_TIM, g_freq_cal_info.need_capture_pulse_sum - 1, 0);
+    tmr_counter_value_set(INPUT_CAPTURE_CRM_TIM, 0);
+    tmr_cnt_dir_set(INPUT_CAPTURE_CRM_TIM, TMR_COUNT_UP);
+
+    g_freq_cal_info.freq_cal_factor = g_freq_cal_info.need_capture_pulse_sum * g_freq_cal_info.cal_capture_pulse_window_size;
+    tmp_pr_value = INPUT_CAPTURE_CAL_CRM_TIM_PR(g_freq_cal_info.freq_cal_timeout_cycle);
+    tmr_base_init(INPUT_CAPTURE_CAL_CRM_TIM, tmp_pr_value, INPUT_CAPTURE_CAL_CRM_DIV -1);
+    tmr_counter_value_set(INPUT_CAPTURE_CAL_CRM_TIM, 0);
+    tmr_cnt_dir_set(INPUT_CAPTURE_CAL_CRM_TIM, TMR_COUNT_UP);
+    tmr_trigger_input_select(INPUT_CAPTURE_CAL_CRM_TIM, TMR_SUB_INPUT_SEL_IS0);
+    tmr_sub_mode_select(INPUT_CAPTURE_CAL_CRM_TIM, TMR_SUB_RESET_MODE);
+    tmr_input_config_struct.input_channel_select = TMR_SELECT_CHANNEL_1;
+    tmr_input_config_struct.input_mapped_select = TMR_CC_CHANNEL_MAPPED_STI;
+    tmr_input_config_struct.input_polarity_select = TMR_INPUT_RISING_EDGE;
+    tmr_input_channel_init(INPUT_CAPTURE_CAL_CRM_TIM, &tmr_input_config_struct, TMR_CHANNEL_INPUT_DIV_1);
+    tmr_external_clock_mode1_config(INPUT_CAPTURE_CRM_TIM, TMR_ES_FREQUENCY_DIV_1, TMR_ES_POLARITY_NON_INVERTED, 0);
+  }
+
+  //interrupt enable
+  tmr_overflow_request_source_set(INPUT_CAPTURE_CAL_CRM_TIM, TRUE);
+  tmr_interrupt_enable(INPUT_CAPTURE_CAL_CRM_TIM, TMR_OVF_INT, TRUE);
+  if (g_freq_cal_info.freq_cal_mode == FIXED_PULSE_NUM_CAL_MODE)
+    tmr_interrupt_enable(INPUT_CAPTURE_CAL_CRM_TIM, TMR_C1_INT, TRUE);
+  armcm_enable_irq(InputCapture_IRQHandler, INPUT_CAPTURE_CAL_IRQn, 0);
+
+  //tmr enable counter
+  tmr_counter_enable(INPUT_CAPTURE_CRM_TIM, TRUE);
+  tmr_counter_enable(INPUT_CAPTURE_CAL_CRM_TIM, TRUE);
+}
+
+void inductance_coil_dev_init(struct freq_cal_info *info) {
+  crm_configuration();
+  gpio_configuration();
+  inductance_coil_crm_tmr_init(info);
+}
+
+#else
+
+#define INPUT_CAPTURE_CAL_COMPENSATION_TICK       (0)
+
+// Todo: adaptation of other chips
+void inductance_coil_dev_init(struct freq_cal_info *info)
+{
+
+}
+
+#endif
+
+uint8_t pulse_gpio_read(void)
+{
+  return g_freq_cal_info.virtual_gpio_state;
+}
+
+uint32_t frequency_conversion(uint32_t tick)
+{
+  uint32_t freq = 0;
+  if (tick) {
+    if (g_freq_cal_info.freq_cal_mode == FIXED_TIME_CAL_MODE) {
+      freq = (uint32_t)(tick * g_freq_cal_info.freq_cal_factor);
+    }
+    else {
+      double single_pulse_tick = 0;
+      single_pulse_tick = (double)(tick + INPUT_CAPTURE_CAL_COMPENSATION_TICK) / g_freq_cal_info.freq_cal_factor;
+      if (single_pulse_tick)
+        freq = (uint32_t)(CONFIG_CLOCK_FREQ / single_pulse_tick);
+    }
+  }
+  return freq;
+}
+
+static uint_fast8_t
+inductance_coil_event(struct timer *t)
+{
+  struct inductance_coil_dev *d = container_of(t, struct inductance_coil_dev, timer);
+
+  sched_wake_task(&inductance_coil_wake);
+
+  // Reschedule timer
+  d->timer.waketime += d->rest_tick;
+  return SF_RESCHEDULE;
+}
+
+void
+command_inductance_coil_config(uint32_t *args)
+{
+  // TODO: currently can only one device exist!!!
+
+  if (inductance_coil_oid_malloc /*&& args[0] != oid */)
+    shutdown("currently only one input_capture resource is supported");
+
+  struct inductance_coil_dev *dev = oid_alloc(args[0], command_inductance_coil_config, sizeof(*dev));
+
+  inductance_coil_oid_malloc = args[0];
+  dev->cal_info.freq_cal_mode = args[1];
+  dev->cal_info.need_capture_pulse_sum = args[2];
+  dev->cal_info.freq_cal_cycle = (double)args[3] / 1000000;
+  dev->cal_info.freq_cal_timeout_cycle = (double)args[4] / 1000000;
+  dev->cal_info.trigger_mode = args[5];
+  dev->cal_info.trigger_invert = args[6];
+  dev->cal_info.trg_freq_ht = args[7];
+  dev->cal_info.trg_freq_lt = args[8];
+  dev->cal_info.cal_capture_pulse_window_size = args[9];
+
+  // initialize peripherals
+  inductance_coil_dev_init(&dev->cal_info);
+
+  dev->timer.func = inductance_coil_event;
+}
+DECL_COMMAND(command_inductance_coil_config,
+            "inductance_coil_config oid=%c cal_mode=%u capture_over_cnt=%u freq_cal_cycle=%u cal_time_out=%u"
+            " trigger_mode=%u trigger_invert=%u trg_freq_ht=%u trg_freq_lt=%u cal_window_size=%u");
+
+static void
+update_trigger_frequencies(struct timer_config_freq_param freq_param)
+{
+  struct inductance_coil_dev *d = oid_lookup(freq_param.oid, command_inductance_coil_config);
+  if (freq_param.absolute_mode) {
+    if (freq_param.trg_freq_ht < 0) freq_param.trg_freq_ht = 0;
+    if (freq_param.trg_freq_lt < 0) freq_param.trg_freq_lt = 0;
+    d->cal_info.trg_freq_ht = freq_param.trg_freq_ht;
+    d->cal_info.trg_freq_lt = freq_param.trg_freq_lt;
+  }
+  else {
+    uint32_t capture_sum = g_freq_cal_info.capture_pulse_sum;
+    uint32_t cur_freq = frequency_conversion(capture_sum);
+
+    if (freq_param.trg_freq_ht + (int)cur_freq < 0)
+      d->cal_info.trg_freq_ht = 0;
+    else
+      d->cal_info.trg_freq_ht = freq_param.trg_freq_ht + cur_freq;
+
+    if (freq_param.trg_freq_lt + (int)cur_freq < 0)
+      d->cal_info.trg_freq_lt = 0;
+    else
+      d->cal_info.trg_freq_lt = freq_param.trg_freq_lt + cur_freq;
+  }
+
+  if (freq_param.trigger_mode == 1 && (d->cal_info.trg_freq_ht < d->cal_info.trg_freq_lt))
+    shutdown("Bidirectional Trigger mode trg_freq_ht is smaller than trg_freq_lt!!!");
+
+  d->cal_info.trigger_mode = freq_param.trigger_mode;
+  d->cal_info.trigger_invert = freq_param.trigger_invert;
+
+  if (freq_param.force_update) {
+    irqstatus_t irq_flags = irq_save();
+    g_freq_cal_info.trg_freq_ht = d->cal_info.trg_freq_ht;
+    g_freq_cal_info.trg_freq_lt = d->cal_info.trg_freq_lt;
+
+    if (g_freq_cal_info.freq_cal_mode == FIXED_TIME_CAL_MODE) {
+      g_freq_cal_info.cmp_trg_cnt_ht = g_freq_cal_info.trg_freq_ht * g_freq_cal_info.freq_cal_cycle * g_freq_cal_info.cal_capture_pulse_window_size;
+      g_freq_cal_info.cmp_trg_cnt_lt = g_freq_cal_info.trg_freq_lt * g_freq_cal_info.freq_cal_cycle * g_freq_cal_info.cal_capture_pulse_window_size;
+    }
+    else {
+      if (g_freq_cal_info.trg_freq_ht) {
+        g_freq_cal_info.cmp_trg_cnt_ht = ((double)INPUT_CAPTURE_CAL_CRM_TIM_CLOCK_FREQ /
+                                            g_freq_cal_info.trg_freq_ht) *  g_freq_cal_info.need_capture_pulse_sum * g_freq_cal_info.cal_capture_pulse_window_size;
+      }
+      else {
+        g_freq_cal_info.cmp_trg_cnt_ht = 0xFFFFFFFF;
+      }
+
+      if (g_freq_cal_info.trg_freq_lt) {
+        g_freq_cal_info.cmp_trg_cnt_lt = ((double)INPUT_CAPTURE_CAL_CRM_TIM_CLOCK_FREQ /
+                                            g_freq_cal_info.trg_freq_lt) *  g_freq_cal_info.need_capture_pulse_sum * g_freq_cal_info.cal_capture_pulse_window_size;
+      }
+      else {
+        g_freq_cal_info.cmp_trg_cnt_lt = 0;
+      }
+    }
+    g_freq_cal_info.trigger_invert = d->cal_info.trigger_invert;
+    g_freq_cal_info.trigger_mode = d->cal_info.trigger_mode;
+    irq_restore(irq_flags);
+  }
+}
+
+static uint_fast8_t
+inductance_coil_config_freq_event(struct timer *t)
+{
+  struct inductance_coil_dev *d = container_of(t, struct inductance_coil_dev, config_freq_timer);
+  update_trigger_frequencies(d->config_freq_param);
+  return SF_DONE;
+}
+
+void
+command_virtual_gpio_trigger_with_timer(uint32_t *args)
+{
+  uint8_t oid = args[0];
+  uint32_t clock = args[7];
+  struct inductance_coil_dev *d = oid_lookup(oid, command_inductance_coil_config);
+
+  irqstatus_t irq_flags = irq_save();
+  sched_del_timer(&d->config_freq_timer);
+
+  d->config_freq_param.oid = oid;
+  d->config_freq_param.absolute_mode = !!args[1];
+  d->config_freq_param.trigger_mode = args[2];
+  d->config_freq_param.trigger_invert = args[3];
+  d->config_freq_param.trg_freq_ht = args[4];
+  d->config_freq_param.trg_freq_lt = args[5];
+  d->config_freq_param.force_update = !!args[6];
+
+  uint32_t current_time = timer_read_time();
+  if (!timer_is_before(current_time, clock)) {
+    update_trigger_frequencies(d->config_freq_param);
+  } else {
+    d->config_freq_timer.func = inductance_coil_config_freq_event;
+    uint32_t min_waketime = current_time + timer_from_us(100);
+    d->config_freq_timer.waketime = timer_is_before(min_waketime, clock) ? clock : min_waketime;
+    sched_add_timer(&d->config_freq_timer);
+  }
+  irq_restore(irq_flags);
+}
+DECL_COMMAND(command_virtual_gpio_trigger_with_timer,
+            "virtual_gpio_trigger_with_timer oid=%c absolute_mode=%u trigger_mode=%u trigger_invert=%u"
+            " trg_freq_ht=%u trg_freq_lt=%u force_update=%u clock=%u");
+
+void
+command_virtual_gpio_trigger(uint32_t *args)
+{
+  struct timer_config_freq_param freq_param;
+  freq_param.oid = args[0];
+  freq_param.absolute_mode = !!args[1];
+  freq_param.trigger_mode = args[2];
+  freq_param.trigger_invert = args[3];
+  freq_param.trg_freq_ht = args[4];
+  freq_param.trg_freq_lt = args[5];
+  freq_param.force_update = !!args[6];
+  update_trigger_frequencies(freq_param);
+}
+DECL_COMMAND(command_virtual_gpio_trigger,
+            "virtual_gpio_trigger oid=%c absolute_mode=%u trigger_mode=%u trigger_invert=%u"
+            " trg_freq_ht=%u trg_freq_lt=%u force_update=%u");
+
+void
+command_inductance_coil_query(uint32_t *args)
+{
+    struct inductance_coil_dev *d = oid_lookup(args[0], command_inductance_coil_config);
+    sched_del_timer(&d->timer);
+    d->timer.waketime = args[1];
+    d->rest_tick = args[2];
+    if (!d->rest_tick)
+        return;
+    sched_add_timer(&d->timer);
+}
+DECL_COMMAND(command_inductance_coil_query,
+             "inductance_coil_query oid=%c clock=%u rest_tick=%u");
+
+void
+command_query_inductance_coil(uint32_t *args)
+{
+  struct inductance_coil_dev *d = oid_lookup(args[0], command_inductance_coil_config);
+  sched_del_timer(&d->timer);
+  if (!args[1])
+      return;
+
+  d->rest_tick = args[1];
+  sensor_bulk_reset(&d->sb);
+
+  irq_disable();
+  d->timer.waketime = timer_read_time() + d->rest_tick;
+  sched_add_timer(&d->timer);
+  irq_enable();
+}
+DECL_COMMAND(command_query_inductance_coil, "query_inductance_coil oid=%c rest_ticks=%u");
+
+void
+command_query_inductance_coil_status(uint32_t *args)
+{
+  struct inductance_coil_dev *icd = oid_lookup(args[0], command_inductance_coil_config);
+
+  irq_disable();
+  uint32_t time = timer_read_time();
+  irq_enable();
+  sensor_bulk_status(&icd->sb, args[0], time, 0, 0);
+}
+DECL_COMMAND(command_query_inductance_coil_status, "query_inductance_coil_status oid=%c");
+
+static void
+inductance_coil_query(struct inductance_coil_dev *icd, uint32_t capture_sum, uint8_t oid)
+{
+  uint32_t cal_freq = 0;
+  cal_freq = frequency_conversion(capture_sum);
+
+  uint8_t *d = &icd->sb.data[icd->sb.data_count];
+  *(uint32_t *)d = cal_freq;
+  icd->sb.data_count += BYTES_PER_SAMPLE;
+
+  // check whether the buffer cannot eat next one sample
+  // if not, must flush the buffer
+  if (icd->sb.data_count + BYTES_PER_SAMPLE > ARRAY_SIZE(icd->sb.data)) {
+    sensor_bulk_report(&icd->sb, oid);
+  }
+}
+
+void
+command_query_inductance_coil_config_info(uint32_t *args)
+{
+  sendf("inductance_coil_info oid=%c cal_mode=%u capture_over_cnt=%u freq_cal_cycle=%u cal_time_out=%u"
+        " trigger_mode=%u trigger_invert=%u trg_freq_ht=%u trg_freq_lt=%u capture_freq=%u virtual_gpio=%u", args[0], g_freq_cal_info.freq_cal_mode,
+        g_freq_cal_info.need_capture_pulse_sum, (uint32_t)(g_freq_cal_info.freq_cal_cycle*1000000), (uint32_t)(g_freq_cal_info.freq_cal_timeout_cycle*1000000),
+        g_freq_cal_info.trigger_mode, g_freq_cal_info.trigger_invert, g_freq_cal_info.trg_freq_ht, g_freq_cal_info.trg_freq_lt,
+        frequency_conversion(g_freq_cal_info.capture_pulse_sum), g_freq_cal_info.virtual_gpio_state);
+}
+DECL_COMMAND(command_query_inductance_coil_config_info, "query_inductance_coil_config_info oid=%c");
+
+void
+inductance_coil_task(void)
+{
+  if (!sched_check_wake(&inductance_coil_wake))
+      return;
+
+  uint8_t oid;
+  struct inductance_coil_dev *dev;
+  foreach_oid(oid, dev, command_inductance_coil_config) {
+    inductance_coil_query(dev, g_freq_cal_info.capture_pulse_sum, oid);
+  }
+}
+DECL_TASK(inductance_coil_task);
+
+void
+inductance_coil_shutdown(void)
+{
+  uint8_t i;
+  struct inductance_coil_dev *d;
+  // Todo: additional specific shutdown handling
+  foreach_oid(i, d, command_inductance_coil_config) {
+    sched_del_timer(&d->timer);
+  }
+}
+DECL_SHUTDOWN(inductance_coil_shutdown);
diff --git a/src/stm32/inductance_coil.h b/src/stm32/inductance_coil.h
new file mode 100644
index 00000000..0815c2ca
--- /dev/null
+++ b/src/stm32/inductance_coil.h
@@ -0,0 +1,6 @@
+#ifndef __INDUCTANCE_COIL_H
+#define __INDUCTANCE_COIL_H
+
+uint8_t pulse_gpio_read(void);
+
+#endif
diff --git a/src/stm32/power_loss_check.c b/src/stm32/power_loss_check.c
new file mode 100644
index 00000000..2521c176
--- /dev/null
+++ b/src/stm32/power_loss_check.c
@@ -0,0 +1,685 @@
+#include <string.h> // ffs
+#include "board/irq.h" // irq_save
+#include "board/misc.h"
+#include "board/armcm_boot.h" // armcm_enable_irq
+#include "command.h" // DECL_ENUMERATION_RANGE
+#include "basecmd.h" // oid_alloc
+#include "sched.h" // sched_shutdown
+#include "autoconf.h" // CONFIG_CLOCK_FREQ
+#include "stepper.h" // get_all_stepper_info
+
+#if CONFIG_MACH_AT32F4x
+// currently it only works with the AT32F415 chip from artery.
+  #if CONFIG_MACH_AT32F415
+    #include "at32f415_crm.h"
+    #include "at32f415_tmr.h"
+    #define FLASH_SECTOR_SIZE                        1024
+    #define RECORD_FLASH_SECTOR_ADDR1                0x0801F800
+    #define RECORD_FLASH_SECTOR_ADDR2                0x0801FC00
+    #define DETECTION_GPIO                           GPIOB
+    #define DETECTION_GPIO_PIN                       GPIO_PINS_8
+  #elif CONFIG_MACH_AT32F403A
+    #include "at32f403a.h"
+    #include "at32f403a_407_crm.h"
+    #include "at32f403a_407_tmr.h"
+    #define FLASH_SECTOR_SIZE                        2048
+    #define RECORD_FLASH_SECTOR_ADDR1                0x080FF000
+    #define RECORD_FLASH_SECTOR_ADDR2                0x080FF800
+    #define DETECTION_GPIO                           GPIOB
+    #define DETECTION_GPIO_PIN                       GPIO_PINS_7
+  #endif
+
+#define TRM_OVER_FREQ          10000
+#define TRM_PR_DIV_VALUE       (uint16_t)(CONFIG_CLOCK_FREQ / TRM_OVER_FREQ - 1)
+#define UNIT16_COUNT(var)      ((size_t)(sizeof(var) / sizeof(uint16_t)))
+
+#define MAX_ALLOW_SAVE_STEPPER_NUM               16            // Number of supported steppers
+#define ENV_VALID_FLAG                           (0x12345678)
+#define ENV_VALID_FLAG_OFFSET                    (0)
+#define ENV_CHECK_SUM_OFFSET                     (4)
+#define ENV_SEQ_NUM_OFFSET                       (8)
+#define ENV_POWER_LOSS_DATA_OFFSET               (12)
+
+struct power_loss_check_dev {
+    struct timer timer;
+    uint8_t report_state;
+    uint8_t need_save;
+    uint8_t print_act;
+    uint8_t voltage_type; // Voltage type: 0-Type1, 1-Type2, 0xFF-Uninitialized
+    uint32_t print_mark;
+    uint32_t report_interval;  // Report interval (microseconds)
+    volatile uint32_t high_level_tick;    // High level tick value (0xFFFFFFFF means not read)
+    volatile uint32_t low_level_tick;     // Low level tick value (0xFFFFFFFF means not read)
+    double duty_cycle_threshold; // Duty cycle threshold (0.0-1.0)
+    double power_loss_trigger_time;
+    uint32_t debounce_count;      // Debounce counter
+    uint32_t debounce_threshold;  // Debounce threshold
+    volatile uint8_t power_loss_flag;      // Power loss flag: 0-Normal, 1-Power loss
+    uint32_t voltage_type_count;  // Continuous recognition counter
+    uint32_t type_confirm_threshold; // Continuous recognition threshold (default 3)
+};
+
+struct __attribute__((aligned(2))) power_loss_env {
+    uint32_t flag;
+    uint16_t step_info_num;
+    struct stepper_info step_info_arry[MAX_ALLOW_SAVE_STEPPER_NUM];
+};
+
+typedef struct {
+    uint32_t addr;
+    uint32_t seq_num;  // Sequence number instead of timestamp
+    uint8_t valid;
+    uint8_t is_init;
+} SectorInfo;
+
+SectorInfo sectors[2] = {
+    {RECORD_FLASH_SECTOR_ADDR1, 0, 0, 0},
+    {RECORD_FLASH_SECTOR_ADDR2, 0, 0, 0}
+};
+
+static uint8_t power_loss_saved = 0;
+static uint8_t save_sector_id = 0;
+static uint32_t save_seq_num = 0;
+static uint32_t last_seq_num = 0;
+static uint32_t valid_count = 0;
+static struct power_loss_check_dev *plc = NULL;
+static struct task_wake power_loss_wake;
+static struct power_loss_env env;
+static struct power_loss_env tmp_env;
+uint16_t flash_buf[FLASH_SECTOR_SIZE / 2];
+
+void power_loss_rotate_sector(void);
+
+/**
+  * @brief  read data using halfword mode
+  * @param  read_addr: the address of reading
+  * @param  p_buffer: the buffer of reading data
+  * @param  num_read: the number of reading data
+  * @retval none
+  */
+void flash_read(uint32_t read_addr, uint8_t *p_buffer, uint32_t num_read)
+{
+    uint32_t i;
+    for(i = 0; i < num_read; i++)
+    {
+        ((uint8_t *)p_buffer)[i] = *((uint8_t*)read_addr);
+        read_addr += 1;
+    }
+}
+
+/**
+  * @brief  write data using halfword mode without checking
+  * @param  write_addr: the address of writing
+  * @param  p_buffer: the buffer of writing data
+  * @param  num_write: the number of writing data
+  * @retval result
+  */
+error_status flash_write_nocheck(uint32_t write_addr, uint16_t *p_buffer, uint16_t num_write)
+{
+    uint16_t i;
+    flash_status_type status = FLASH_OPERATE_DONE;
+    for(i = 0; i < num_write; i++)
+    {
+        status = flash_halfword_program(write_addr, p_buffer[i]);
+        if(status != FLASH_OPERATE_DONE)
+            return ERROR;
+        write_addr += 2;
+    }
+    return SUCCESS;
+}
+
+  /**
+    * @brief  write data using halfword mode with checking
+    * @param  write_addr: the address of writing
+    * @param  p_buffer: the buffer of writing data
+    * @param  num_write: the number of writing data
+    * @retval result
+    */
+error_status flash_write(uint32_t write_addr, uint16_t *p_buffer, uint16_t num_write)
+{
+    uint32_t offset_addr;
+    uint32_t sector_position;
+    uint16_t sector_offset;
+    uint16_t sector_remain;
+    uint16_t i;
+    flash_status_type status = FLASH_OPERATE_DONE;
+    flash_unlock();
+    offset_addr = write_addr - FLASH_BASE;
+    sector_position = offset_addr / FLASH_SECTOR_SIZE;
+    sector_offset = (offset_addr % FLASH_SECTOR_SIZE) / 2;
+    sector_remain = FLASH_SECTOR_SIZE / 2 - sector_offset;
+    if(num_write <= sector_remain)
+        sector_remain = num_write;
+    while(1)
+    {
+        flash_read(sector_position * FLASH_SECTOR_SIZE + FLASH_BASE, (uint8_t *)flash_buf, FLASH_SECTOR_SIZE);
+        for(i = 0; i < sector_remain; i++)
+        {
+            if(flash_buf[sector_offset + i] != 0xFFFF)
+                break;
+        }
+
+        if(i < sector_remain)
+        {
+            /* wait for operation to be completed */
+            status = flash_operation_wait_for(ERASE_TIMEOUT);
+            if((status == FLASH_PROGRAM_ERROR) || (status == FLASH_EPP_ERROR))
+                flash_flag_clear(FLASH_PRGMERR_FLAG | FLASH_EPPERR_FLAG);
+            else if(status == FLASH_OPERATE_TIMEOUT)
+                return ERROR;
+            status = flash_sector_erase(sector_position * FLASH_SECTOR_SIZE + FLASH_BASE);
+            if(status != FLASH_OPERATE_DONE)
+                return ERROR;
+            for(i = 0; i < sector_remain; i++)
+            {
+                flash_buf[i + sector_offset] = p_buffer[i];
+            }
+            if(flash_write_nocheck(sector_position * FLASH_SECTOR_SIZE + FLASH_BASE, flash_buf, FLASH_SECTOR_SIZE / 2) != SUCCESS)
+                return ERROR;
+        }
+        else
+        {
+            if(flash_write_nocheck(write_addr, p_buffer, sector_remain) != SUCCESS)
+                return ERROR;
+        }
+
+        if(num_write == sector_remain)
+            break;
+        else
+        {
+            sector_position++;
+            sector_offset = 0;
+            p_buffer += sector_remain;
+            write_addr += (sector_remain * 2);
+            num_write -= sector_remain;
+            if(num_write > (FLASH_SECTOR_SIZE / 2))
+            sector_remain = FLASH_SECTOR_SIZE / 2;
+            else
+            sector_remain = num_write;
+        }
+    }
+    flash_lock();
+    return SUCCESS;
+  }
+
+void flash_sector_erase_ex(uint32_t sector_address) {
+    flash_unlock();
+    flash_sector_erase(sector_address);
+    flash_lock();
+}
+
+uint16_t cal_checksum(const void* data, uint32_t len) {
+    const uint8_t* addr = (const uint8_t*)data;
+    uint32_t sum = 0;
+
+    while (len > 1) {
+        sum += (addr[0] << 8) | addr[1];
+        addr += 2;
+        len -= 2;
+    }
+
+    if (len == 1) {
+        sum += (uint16_t)(*(const uint8_t*)addr) << 8;
+    }
+
+    sum = (sum >> 16) + (sum & 0xFFFF);
+    sum += (sum >> 16);
+
+    return (uint16_t)~sum;
+}
+
+uint8_t is_all_value(const uint8_t* data, uint8_t value, size_t len) {
+    const uint8_t* end = data + len;
+    while (data < end) {
+        if (*data++ != value)
+            return 0xFF;
+    }
+    return 0;
+}
+
+void power_loss_rotate_sector(void)
+{
+    if (plc) {
+        uint32_t checksum = 0;
+        uint16_t step_num = 0;
+        memset((uint8_t*)flash_buf, 0, FLASH_SECTOR_SIZE);
+        // *((uint8_t*)flash_buf + ENV_SEQ_NUM_OFFFSET) = save_seq_num;
+        memcpy((uint8_t*)flash_buf + ENV_SEQ_NUM_OFFSET, &save_seq_num, sizeof(uint32_t));
+        step_num = get_all_stepper_info(env.step_info_arry, MAX_ALLOW_SAVE_STEPPER_NUM, 1);
+        env.flag = plc->print_mark;
+        env.step_info_num = step_num;
+        memcpy(((uint8_t*)flash_buf)+ENV_POWER_LOSS_DATA_OFFSET, &env, sizeof(env));
+        checksum = cal_checksum(((uint8_t*)flash_buf)+ENV_SEQ_NUM_OFFSET, sizeof(env)+4);
+        flash_write(sectors[save_sector_id].addr + ENV_CHECK_SUM_OFFSET, (uint16_t*)&checksum, UNIT16_COUNT(checksum));
+        flash_write(sectors[save_sector_id].addr + ENV_SEQ_NUM_OFFSET, (uint16_t*)&save_seq_num, UNIT16_COUNT(save_seq_num));
+        flash_write(sectors[save_sector_id].addr + ENV_POWER_LOSS_DATA_OFFSET, (uint16_t*)&env, UNIT16_COUNT(env));
+    }
+}
+
+void load_save_flash_info(void)
+{
+    uint32_t checksum = 0;
+    uint32_t read_check_sum = 0;
+    uint32_t seq = 0;
+    uint32_t env_valid_flag = ENV_VALID_FLAG;
+
+    for (int i = 0; i < 2; i++) {
+        memset((uint8_t*)flash_buf, 0, FLASH_SECTOR_SIZE);
+        flash_read(sectors[i].addr, (uint8_t*)flash_buf, FLASH_SECTOR_SIZE);
+        if (*((uint32_t*)flash_buf) == ENV_VALID_FLAG) {
+            if (is_all_value(((uint8_t*)flash_buf)+ENV_CHECK_SUM_OFFSET, 0xFF, sizeof(struct power_loss_env)+8) == 0) {
+                sectors[i].is_init = 1;
+            }
+            else {
+                read_check_sum = *(((uint32_t*)flash_buf) + 1);
+                checksum = cal_checksum(((uint8_t*)flash_buf)+ENV_SEQ_NUM_OFFSET, sizeof(struct power_loss_env)+4);
+                seq = *(((uint32_t*)flash_buf) + 2);
+                if (read_check_sum == checksum) {
+                    if (seq != 0xFFFFFFFF && seq != 0x0) {
+                        sectors[i].valid = 1;
+                        sectors[i].seq_num = seq;
+                        valid_count++;
+                    }
+                }
+            }
+        }
+    }
+
+    if (valid_count == 2) {
+        // Special index handling
+        if ((sectors[0].seq_num == 1 && sectors[1].seq_num == 0xFFFFFFFE) || ((sectors[0].seq_num == 0xFFFFFFFE && sectors[1].seq_num == 1))) {
+            save_sector_id = (sectors[0].seq_num == 1) ? 1 : 0;
+            last_seq_num = 1;
+        }
+        else {
+            save_sector_id = (sectors[0].seq_num < sectors[1].seq_num) ? 0 : 1;
+            last_seq_num = (sectors[0].seq_num < sectors[1].seq_num) ? (sectors[1].seq_num) : (sectors[0].seq_num);
+        }
+        save_seq_num = last_seq_num + 1;
+        if (save_seq_num == 0xFFFFFFFF || save_seq_num == 0)
+            save_seq_num = 1;
+        if (!sectors[save_sector_id].is_init) {
+            flash_sector_erase_ex(sectors[save_sector_id].addr);
+            flash_write(sectors[save_sector_id].addr, (uint16_t*)&env_valid_flag, UNIT16_COUNT(env_valid_flag));
+        }
+    }
+    else if (valid_count == 1) {
+        save_sector_id = sectors[0].valid ? 1 : 0;
+        last_seq_num = sectors[!save_sector_id].seq_num;
+        save_seq_num = last_seq_num + 1;
+        if (save_seq_num == 0xFFFFFFFF || save_seq_num == 0)
+            save_seq_num = 1;
+        if (!sectors[save_sector_id].is_init) {
+            flash_sector_erase_ex(sectors[save_sector_id].addr);
+            flash_write(sectors[save_sector_id].addr, (uint16_t*)&env_valid_flag, UNIT16_COUNT(env_valid_flag));
+        }
+    }
+    else {
+        for (int i = 0; i < 2; i++) {
+            if (!sectors[i].is_init) {
+                flash_sector_erase_ex(sectors[i].addr);
+                flash_write(sectors[i].addr, (uint16_t*)&env_valid_flag, UNIT16_COUNT(env_valid_flag));
+            }
+        }
+        save_sector_id = 0;
+        save_seq_num = 1;
+    }
+    memset((uint8_t*)(&env), 0, sizeof(env));
+    if (valid_count) {
+        uint8_t last_valid_sector = save_sector_id ? 0: 1;
+        flash_read(sectors[last_valid_sector].addr+ENV_POWER_LOSS_DATA_OFFSET, (uint8_t*)(&env), sizeof(env));
+    }
+}
+
+void
+trigger_power_loss_gpio_set(void)
+{
+    #if CONFIG_MACH_AT32F403A
+        GPIOC->scr = GPIO_PINS_8;
+        GPIOC->scr = GPIO_PINS_9;
+        GPIOD->clr = GPIO_PINS_4;
+        GPIOD->clr = GPIO_PINS_5;
+        GPIOD->clr = GPIO_PINS_6;
+        GPIOD->clr = GPIO_PINS_7;
+        GPIOE->clr = GPIO_PINS_15;
+    #else
+        GPIOB->clr = GPIO_PINS_3;
+        GPIOB->scr = GPIO_PINS_0;
+        GPIOB->scr = GPIO_PINS_7;
+    #endif
+}
+
+void
+#if CONFIG_MACH_AT32F403A
+TIM4_IRQHandler(void)
+{
+    uint32_t ists = TMR4->ists;
+    if (plc) {
+        if (plc->debounce_count >= plc->debounce_threshold) {
+            if (ists & TMR_C2_FLAG) {
+                const uint32_t current_tick = TMR4->cval;
+                const uint8_t current_level = gpio_input_data_bit_read(DETECTION_GPIO, DETECTION_GPIO_PIN);
+                if (current_level) {
+                    plc->low_level_tick = current_tick;
+                    if (plc->high_level_tick != 0xFFFFFFFF) {
+                        uint32_t period = plc->high_level_tick + plc->low_level_tick;
+                        uint32_t threshold_tick = (uint32_t)(period * plc->duty_cycle_threshold);
+                        uint8_t new_type = (plc->high_level_tick > threshold_tick) ? 1 : 0;
+
+                        if (plc->voltage_type == new_type) {
+                            plc->voltage_type_count++;
+                        } else {
+                            plc->voltage_type_count = 0;
+                            plc->voltage_type = new_type;
+                        }
+
+                        if (plc->voltage_type_count >= plc->type_confirm_threshold) {
+                            plc->voltage_type = new_type;
+                        }
+                    }
+                } else {
+                    plc->high_level_tick = current_tick;
+                }
+            }
+
+            if (ists & TMR_OVF_FLAG) {
+                if (plc->power_loss_flag != 1 && plc->print_act && !power_loss_saved) {
+                    GPIOA->scr = GPIO_PINS_15;
+                    plc->need_save = 1;
+                    plc->power_loss_flag = 1;
+                    sched_wake_task(&power_loss_wake);
+                }
+            }
+
+            if (plc->power_loss_flag == 1) {
+                trigger_power_loss_gpio_set();
+            }
+        }
+        else {
+            plc->debounce_count++;
+            plc->power_loss_flag = 0;
+        }
+    }
+
+    if (ists & (TMR_C2_FLAG | TMR_OVF_FLAG)) {
+        TMR4->ists = ~(ists & (TMR_C2_FLAG | TMR_OVF_FLAG));
+        TMR4->cval = 0;
+    }
+}
+#else
+TIM4_IRQHandler(void)
+{
+    uint32_t ists = TMR4->ists;
+    if (plc) {
+        if (plc->debounce_count >= plc->debounce_threshold) {
+            if (ists & TMR_OVF_FLAG) {
+                const uint8_t current_level = gpio_input_data_bit_read(DETECTION_GPIO, DETECTION_GPIO_PIN);
+                if (current_level) {
+                    plc->voltage_type_count++;
+                    if (plc->voltage_type_count >= plc->type_confirm_threshold) {
+                        if (plc->power_loss_flag != 1 && plc->print_act && !power_loss_saved) {
+                            plc->need_save = 1;
+                            plc->power_loss_flag = 1;
+                            sched_wake_task(&power_loss_wake);
+                        }
+                    }
+                } else {
+                    plc->voltage_type_count = 0;
+                }
+            }
+
+            if (plc->power_loss_flag == 1) {
+                trigger_power_loss_gpio_set();
+            }
+        }
+        else {
+            plc->debounce_count++;
+            plc->power_loss_flag = 0;
+            plc->voltage_type_count = 0;
+        }
+    }
+    TMR4->ists = ~TMR_OVF_FLAG;
+    TMR4->cval = 0;
+}
+#endif
+
+static uint_fast8_t
+power_loss_report_event(struct timer *t)
+{
+    struct power_loss_check_dev *p = container_of(t, struct power_loss_check_dev, timer);
+    p->report_state = 1;
+    sched_wake_task(&power_loss_wake);
+    p->timer.waketime += p->report_interval;
+    return SF_RESCHEDULE;
+}
+
+void
+power_loss_check_init(void)
+{
+    if (!plc) {
+        output("power_loss_check_dev: Missing config");
+        return;
+    }
+    crm_periph_clock_enable(CRM_TMR4_PERIPH_CLOCK, TRUE);
+    crm_periph_clock_enable(CRM_GPIOB_PERIPH_CLOCK, TRUE);
+    crm_periph_clock_enable(CRM_GPIOA_PERIPH_CLOCK, TRUE);
+
+    // gpio init
+    gpio_init_type gpio_init_struct = {0};
+    gpio_init_struct.gpio_pins = DETECTION_GPIO_PIN;
+    gpio_init_struct.gpio_mode = GPIO_MODE_INPUT;
+    gpio_init_struct.gpio_out_type = GPIO_OUTPUT_PUSH_PULL;
+    gpio_init_struct.gpio_pull = GPIO_PULL_DOWN;
+    gpio_init_struct.gpio_drive_strength = GPIO_DRIVE_STRENGTH_STRONGER;
+    gpio_init(DETECTION_GPIO, &gpio_init_struct);
+
+    #if CONFIG_MACH_AT32F403A
+        gpio_init_struct.gpio_drive_strength = GPIO_DRIVE_STRENGTH_STRONGER;
+        gpio_init_struct.gpio_out_type = GPIO_OUTPUT_PUSH_PULL;
+        gpio_init_struct.gpio_mode = GPIO_MODE_OUTPUT;
+        gpio_init_struct.gpio_pins = GPIO_PINS_15;
+        gpio_init_struct.gpio_pull = GPIO_PULL_NONE;
+        gpio_init(GPIOA, &gpio_init_struct);
+        GPIOA->clr = GPIO_PINS_15;
+    #endif
+
+    // tmr init
+    tmr_counter_enable(TMR4, FALSE);
+    tmr_interrupt_enable(TMR4, TMR_OVF_INT | TMR_C2_INT, FALSE);
+    TMR4->ists &= 0;
+    /* tmr4 counter mode configuration */
+    tmr_base_init(TMR4, 0xFFFF, TRM_PR_DIV_VALUE);
+
+    tmr_base_init(TMR4, (uint16_t)(TRM_OVER_FREQ*plc->power_loss_trigger_time - 1), TRM_PR_DIV_VALUE);
+    tmr_counter_value_set(TMR4, 0);
+    tmr_cnt_dir_set(TMR4, TMR_COUNT_UP);
+
+    #if CONFIG_MACH_AT32F403A
+        tmr_input_config_type tmr_input_config_struct;
+        tmr_input_config_struct.input_channel_select = TMR_SELECT_CHANNEL_2;
+        tmr_input_config_struct.input_mapped_select = TMR_CC_CHANNEL_MAPPED_DIRECT;
+        tmr_input_config_struct.input_polarity_select = TMR_INPUT_BOTH_EDGE;
+        tmr_input_channel_init(TMR4, &tmr_input_config_struct, TMR_CHANNEL_INPUT_DIV_1);
+        tmr_interrupt_enable(TMR4, TMR_C2_INT, TRUE);
+    #endif
+
+    tmr_overflow_request_source_set(TMR4, TRUE);
+    tmr_interrupt_enable(TMR4, TMR_OVF_INT, TRUE);
+    armcm_enable_irq(TIM4_IRQHandler, TMR4_GLOBAL_IRQn, 1);
+
+    // tmr enable counter
+    tmr_counter_enable(TMR4, TRUE);
+}
+
+void
+command_config_power_loss_check_dev(uint32_t *args)
+{
+    if (plc) {
+        output("power_loss_check_dev: Already configured");
+        return;
+    }
+    plc = oid_alloc(args[0], command_config_power_loss_check_dev, sizeof(*plc));
+    plc->timer.waketime = args[1];
+    plc->power_loss_trigger_time = (double)args[2] / 1000000;
+    plc->report_interval = args[3];
+    plc->duty_cycle_threshold = (double)args[4] / 1000000;
+    plc->debounce_threshold = args[5];
+    plc->report_state = 0;
+    plc->need_save = 0;
+    plc->high_level_tick = 0xFFFFFFFF;
+    plc->low_level_tick = 0xFFFFFFFF;
+    plc->voltage_type = 0xFF;
+    plc->timer.func = power_loss_report_event;
+    plc->debounce_count = 0;
+    plc->power_loss_flag = 0;
+    plc->type_confirm_threshold = args[6] ? args[6] : 3;
+    plc->print_act = 0;
+    plc->print_mark = 0xFFFFFFFF;
+    power_loss_check_init();
+    if (plc->report_interval)
+      sched_add_timer(&plc->timer);
+}
+DECL_COMMAND(command_config_power_loss_check_dev,
+             "config_power_loss_check oid=%c clock=%u power_loss_trigger_time=%u report_interval=%u duty_threshold=%u"
+             " debounce_threshold=%u type_confirm_threshold=%u");
+
+void
+command_query_power_loss_status(uint32_t *args)
+{
+    uint32_t high_level = plc ? plc->high_level_tick : 0xFFFFFFFF;
+    uint32_t low_level = plc ? plc->low_level_tick : 0xFFFFFFFF;
+    uint8_t voltage_type = plc ? plc->voltage_type : 0xFF;
+    uint8_t power_loss_flag = plc ? plc->power_loss_flag : 0;
+    uint8_t initialized = plc ? 1 : 0;
+
+    sendf("power_loss_status oid=%c high_level=%u low_level=%u voltage_type=%u power_loss_flag=%u initialized=%u",
+          args[0], high_level, low_level, voltage_type, power_loss_flag, initialized);
+}
+DECL_COMMAND(command_query_power_loss_status, "query_power_loss_status oid=%c");
+
+void
+command_update_report_interval(uint32_t *args)
+{
+    if (!plc) {
+        output("power_loss_check_dev: Not configured");
+        return;
+    }
+
+    irq_disable();
+    sched_del_timer(&plc->timer);
+    plc->report_interval = args[1];
+    if (plc->report_interval) {
+        plc->timer.waketime = timer_read_time() + plc->report_interval;
+        sched_add_timer(&plc->timer);
+    }
+    irq_enable();
+}
+DECL_COMMAND(command_update_report_interval,
+             "update_report_interval oid=%c report_interval=%u");
+
+void
+command_enable_power_loss(uint32_t *args)
+{
+    if (!plc) {
+        output("power_loss_check_dev: Not configured");
+        return;
+    }
+
+    irq_disable();
+    plc->print_act = !!args[1];
+    if (plc->print_act)
+        plc->print_mark = args[2];
+    else
+        plc->print_mark = 0xFFFFFFFF;
+    config_stepper_print_act(!!args[1], args[3]);
+    irq_enable();
+}
+DECL_COMMAND(command_enable_power_loss,
+            "enable_power_loss oid=%c enable=%u print_flag=%u move_line=%u");
+
+void
+command_query_power_loss_flash_valid(uint32_t *args)
+{
+    sendf("power_loss_flash_valid oid=%c last_seq=%u valid_sector_count=%u env_flag=%u save_stepper_num=%u",
+        args[0], last_seq_num, valid_count, env.flag, env.step_info_num);
+}
+DECL_COMMAND(command_query_power_loss_flash_valid, "query_power_loss_flash_valid oid=%c");
+
+void
+command_query_power_loss_stepper_info(uint32_t *args)
+{
+    uint16_t step_info_num = env.step_info_num;
+    uint16_t i = 0;
+    if (args[1] == 0xFF) {
+        for (i = 0; i < step_info_num && i < MAX_ALLOW_SAVE_STEPPER_NUM; i++) {
+            sendf("power_loss_stepper_info oid=%c type=%u index=%u line=%u position=%u",
+                args[0], env.step_info_arry[i].type, env.step_info_arry[i].index,
+                env.step_info_arry[i].line, env.step_info_arry[i].position);
+        }
+    }
+    else {
+        step_info_num = get_all_stepper_info(tmp_env.step_info_arry, MAX_ALLOW_SAVE_STEPPER_NUM, 0);
+        for (i = 0; i < step_info_num && i < MAX_ALLOW_SAVE_STEPPER_NUM; i++) {
+            sendf("power_loss_stepper_info oid=%c type=%u index=%u line=%u position=%u",
+                args[0], tmp_env.step_info_arry[i].type, tmp_env.step_info_arry[i].index,
+                tmp_env.step_info_arry[i].line, tmp_env.step_info_arry[i].position);
+        }
+    }
+    sendf("power_loss_stepper_info_result oid=%c result=%u", args[0], 1);
+}
+DECL_COMMAND(command_query_power_loss_stepper_info, "query_power_loss_stepper_info oid=%c type=%u index=%u");
+
+void
+power_loss_check_dev_task(void)
+{
+    if (plc && plc->power_loss_flag == 1)
+        trigger_power_loss_gpio_set();
+
+    if (!sched_check_wake(&power_loss_wake))
+        return;
+
+    uint8_t oid;
+    struct power_loss_check_dev *p;
+    foreach_oid(oid, p, command_config_power_loss_check_dev) {
+        if (plc->need_save && !power_loss_saved) {
+            power_loss_saved = 1;
+            power_loss_rotate_sector();
+            output("Power loss info saved");
+            plc->need_save = 0;
+        }
+
+        irq_disable();
+        if (!p->report_state) {
+            irq_enable();
+            continue;
+        }
+        p->report_state = 0;
+        irq_enable();
+        uint8_t initialized = plc ? 1 : 0;
+        sendf("report_power_loss_status oid=%c high_level=%u low_level=%u voltage_type=%u power_loss_flag=%u initialized=%u",
+              oid, plc->high_level_tick, plc->low_level_tick, plc->voltage_type, plc->power_loss_flag, initialized);
+    }
+}
+DECL_TASK(power_loss_check_dev_task);
+
+void power_loss_check_task_init(void)
+{
+    load_save_flash_info();
+}
+DECL_INIT(power_loss_check_task_init);
+
+void
+power_loss_check_shutdown(void)
+{
+    uint8_t oid;
+    struct power_loss_check_dev *p;
+    foreach_oid(oid, p, command_config_power_loss_check_dev) {
+        if (!power_loss_saved && plc->print_act) {
+            power_loss_saved = 1;
+            power_loss_rotate_sector();
+            output("Power loss info saved during shutdown");
+        }
+    }
+}
+DECL_SHUTDOWN(power_loss_check_shutdown);
+#endif
diff --git a/src/stm32/serial.c b/src/stm32/serial.c
index 19aa048e..5975db9c 100644
--- a/src/stm32/serial.c
+++ b/src/stm32/serial.c
@@ -48,6 +48,12 @@
   #define GPIO_Tx GPIO('D', 8)
   #define USARTx USART3
   #define USARTx_IRQn USART3_IRQn
+#elif CONFIG_STM32_SERIAL_AT_USART5_PB8_PB9 || CONFIG_STM32_USBCANBUS_PA11_PA12_AND_SERIAL_USART5_PB8_PB9
+  DECL_CONSTANT_STR("RESERVE_PINS_serial", "PB9,PB8");
+  #define GPIO_Rx GPIO('B', 8)
+  #define GPIO_Tx GPIO('B', 9)
+  #define USARTx UART5
+  #define USARTx_IRQn UART5_IRQn
 #endif
 
 #define CR1_FLAGS (USART_CR1_UE | USART_CR1_RE | USART_CR1_TE   \
@@ -92,5 +98,10 @@ serial_init(void)
 
     gpio_peripheral(GPIO_Rx, GPIO_FUNCTION(7), 1);
     gpio_peripheral(GPIO_Tx, GPIO_FUNCTION(7), 0);
+
+  #if CONFIG_STM32_SERIAL_AT_USART5_PB8_PB9 || CONFIG_STM32_USBCANBUS_PA11_PA12_AND_SERIAL_USART5_PB8_PB9
+    extern void mcu_uart_gpio_remap();
+    mcu_uart_gpio_remap();
+  #endif
 }
 DECL_INIT(serial_init);
diff --git a/src/stm32/spi.c b/src/stm32/spi.c
index 648e9bcd..90fb2891 100644
--- a/src/stm32/spi.c
+++ b/src/stm32/spi.c
@@ -32,15 +32,22 @@ DECL_CONSTANT_STR("BUS_PINS_spi1a", "PB4,PB5,PB3");
 #ifdef SPI3
  DECL_ENUMERATION("spi_bus", "spi3", 4);
  DECL_CONSTANT_STR("BUS_PINS_spi3", "PB4,PB5,PB3");
- #if CONFIG_MACH_STM32F4 || CONFIG_MACH_STM32G4
+ #if CONFIG_MACH_STM32F4 || CONFIG_MACH_STM32G4 || CONFIG_MACH_AT32F403A
   DECL_ENUMERATION("spi_bus", "spi3a", 5);
   DECL_CONSTANT_STR("BUS_PINS_spi3a", "PC11,PC12,PC10");
-  #ifdef SPI4
-   DECL_ENUMERATION("spi_bus", "spi4", 6);
-   DECL_CONSTANT_STR("BUS_PINS_spi4", "PE13,PE14,PE12");
-  #elif defined(GPIOI)
-   DECL_ENUMERATION("spi_bus", "spi2b", 6);
-   DECL_CONSTANT_STR("BUS_PINS_spi2b", "PI2,PI3,PI1");
+  #if CONFIG_MACH_AT32F403A
+    #define SPI4_BASE             (APB1PERIPH_BASE + 0x00004000UL)
+    #define SPI4                ((SPI_TypeDef *)SPI4_BASE)
+    DECL_ENUMERATION("spi_bus", "spi4", 6);
+    DECL_CONSTANT_STR("BUS_PINS_spi4", "PE13,PE14,PE11");
+  #else
+    #ifdef SPI4
+      DECL_ENUMERATION("spi_bus", "spi4", 6);
+      DECL_CONSTANT_STR("BUS_PINS_spi4", "PE13,PE14,PE12");
+    #elif defined(GPIOI)
+      DECL_ENUMERATION("spi_bus", "spi2b", 6);
+      DECL_CONSTANT_STR("BUS_PINS_spi2b", "PI2,PI3,PI1");
+    #endif
   #endif
  #endif
  #if CONFIG_MACH_STM32G0B1
@@ -62,17 +69,21 @@ static const struct spi_info spi_bus[] = {
     { SPI2, GPIO('B', 14), GPIO('B', 15), GPIO('B', 13), SPI_FUNCTION_ALL },
     { SPI1, GPIO('A', 6), GPIO('A', 7), GPIO('A', 5), SPI_FUNCTION_ALL },
     { SPI1, GPIO('B', 4), GPIO('B', 5), GPIO('B', 3), SPI_FUNCTION_ALL },
-#if CONFIG_MACH_STM32G4
+#if CONFIG_MACH_STM32G4 || CONFIG_MACH_AT32F403A
     { SPI2, GPIO('A', 10), GPIO('A', 11), GPIO('F', 1), SPI_FUNCTION_ALL },
 #else
     { SPI2, GPIO('C', 2), GPIO('C', 3), GPIO('B', 10), SPI_FUNCTION_ALL },
 #endif
 #ifdef SPI3
     { SPI3, GPIO('B', 4), GPIO('B', 5), GPIO('B', 3), GPIO_FUNCTION_ALL(6) },
- #if CONFIG_MACH_STM32F4 || CONFIG_MACH_STM32G4
+ #if CONFIG_MACH_STM32F4 || CONFIG_MACH_STM32G4 || CONFIG_MACH_AT32F403A
     { SPI3, GPIO('C', 11), GPIO('C', 12), GPIO('C', 10), GPIO_FUNCTION_ALL(6) },
   #ifdef SPI4
-    { SPI4, GPIO('E', 13), GPIO('E', 14), GPIO('E', 12), GPIO_FUNCTION_ALL(5) },
+    #if !CONFIG_MACH_AT32F403A
+        { SPI4, GPIO('E', 13), GPIO('E', 14), GPIO('E', 12), GPIO_FUNCTION_ALL(5) },
+    #else
+        { SPI4, GPIO('E', 13), GPIO('E', 14), GPIO('E', 11), GPIO_FUNCTION_ALL(5) },
+    #endif
   #elif defined(GPIOI)
     { SPI2, GPIO('I', 2), GPIO('I', 3), GPIO('I', 1), GPIO_FUNCTION_ALL(5) },
   #endif
@@ -94,6 +105,11 @@ spi_setup(uint32_t bus, uint8_t mode, uint32_t rate)
     SPI_TypeDef *spi = spi_bus[bus].spi;
     if (!is_enabled_pclock((uint32_t)spi)) {
         enable_pclock((uint32_t)spi);
+        #if CONFIG_MACH_AT32F403A
+            extern void mcu_spi4_gpio_remap(void);
+            if (spi_bus[bus].miso_pin == GPIO('E', 13))
+                mcu_spi4_gpio_remap();
+        #endif
         gpio_peripheral(spi_bus[bus].miso_pin, spi_bus[bus].miso_af, 1);
         gpio_peripheral(spi_bus[bus].mosi_pin, spi_bus[bus].mosi_af, 0);
         gpio_peripheral(spi_bus[bus].sck_pin, spi_bus[bus].sck_af, 0);
@@ -126,6 +142,7 @@ spi_prepare(struct spi_config config)
         return;
     // The SPE bit must be disabled before changing CPOL/CPHA bits
     spi->CR1 = cr1 & ~SPI_CR1_SPE;
+    spi->CR1 = config.spi_cr1 & ~SPI_CR1_SPE; // Modify CPOL/CPHA bits in SPE bit disabled state
     spi->CR1; // Force flush of previous write
     spi->CR1 = config.spi_cr1;
 }
diff --git a/src/stm32/stm32f1.c b/src/stm32/stm32f1.c
index 9d3c3b97..1d106d8f 100644
--- a/src/stm32/stm32f1.c
+++ b/src/stm32/stm32f1.c
@@ -51,6 +51,11 @@ gpio_clock_enable(GPIO_TypeDef *regs)
     RCC->APB2ENR;
 }
 
+#if CONFIG_MACH_AT32F415
+    #include "at32f415rc.h"
+#elif CONFIG_MACH_AT32F403A
+    #include "at32f403a.h"
+#else
 // Main clock setup called at chip startup
 static void
 clock_setup(void)
@@ -89,7 +94,7 @@ clock_setup(void)
     while ((RCC->CFGR & RCC_CFGR_SWS_Msk) != RCC_CFGR_SWS_PLL)
         ;
 }
-
+#endif
 
 /****************************************************************
  * GPIO setup
@@ -270,7 +275,13 @@ armcm_main(void)
     RCC->APB2ENR = 0;
 
     // Setup clocks
+#if CONFIG_MACH_AT32F415
+    at32f415rc_clock_setup();
+#elif CONFIG_MACH_AT32F403A
+    at32f403a_clock_setup();
+#else
     clock_setup();
+#endif
 
     // Disable JTAG to free PA15, PB3, PB4
     enable_pclock(AFIO_BASE);
@@ -282,5 +293,10 @@ armcm_main(void)
         stm32f1_alternative_remap(AFIO_MAPR_SWJ_CFG_Msk,
                                   AFIO_MAPR_SWJ_CFG_JTAGDISABLE);
 
+#if CONFIG_MACH_AT32F4x && CONFIG_AT32_ENABLE_DEBUG_USART
+    uart_debug_print_init(460800);
+    LOG_I("mcu start up\n");
+#endif
+
     sched_main();
 }
diff --git a/src/stm32/usbfs.c b/src/stm32/usbfs.c
index 5385c956..8bddbba3 100644
--- a/src/stm32/usbfs.c
+++ b/src/stm32/usbfs.c
@@ -265,7 +265,8 @@ usb_send_bulk_in(void *data, uint_fast8_t len)
             // buffering mode, so wait for second packet before starting.
             if (bipp == (BI_START | 1)) {
                 bulk_in_push_pos = 0;
-                writel(&bulk_in_pop_flag, USB_EP_KIND); // Dummy flag
+                if (!CONFIG_MACH_AT32F403A)
+                    writel(&bulk_in_pop_flag, USB_EP_KIND); // Dummy flag
                 USB_EPR[ep] = calc_epr_bits(epr, USB_EPTX_STAT
                                             , USB_EP_TX_VALID);
             }
diff --git a/src/stm32/usbotg.c b/src/stm32/usbotg.c
index b2d52456..e46b27ee 100644
--- a/src/stm32/usbotg.c
+++ b/src/stm32/usbotg.c
@@ -14,22 +14,35 @@
 #include "internal.h" // GPIO
 #include "sched.h" // DECL_INIT
 
-#if CONFIG_STM32_USB_PB14_PB15
-  #define IS_OTG_HS 1
-  #define GPIO_D_NEG GPIO('B', 14)
-  #define GPIO_D_POS GPIO('B', 15)
-  #define GPIO_FUNC GPIO_FUNCTION(12)
-  DECL_CONSTANT_STR("RESERVE_PINS_USB1", "PB14,PB15");
+#if !CONFIG_MACH_AT32F415
+    #if CONFIG_STM32_USB_PB14_PB15
+        #define IS_OTG_HS 1
+        #define GPIO_D_NEG GPIO('B', 14)
+        #define GPIO_D_POS GPIO('B', 15)
+        #define GPIO_FUNC GPIO_FUNCTION(12)
+        DECL_CONSTANT_STR("RESERVE_PINS_USB1", "PB14,PB15");
+    #else
+        #if CONFIG_MACH_STM32H723
+            #define IS_OTG_HS 1
+        #else
+            #define IS_OTG_HS 0
+        #endif
+        #define GPIO_D_NEG GPIO('A', 11)
+        #define GPIO_D_POS GPIO('A', 12)
+        #define GPIO_FUNC GPIO_FUNCTION(10)
+        DECL_CONSTANT_STR("RESERVE_PINS_USB", "PA11,PA12");
+    #endif
 #else
-  #if CONFIG_MACH_STM32H723
-    #define IS_OTG_HS 1
-  #else
+    #include "at32f415rc.h"
+    #define USB_OTG_GCCFG_NOVBUSSENS_Pos             (21U)
+    #define USB_OTG_GCCFG_NOVBUSSENS_Msk             (0x1UL << USB_OTG_GCCFG_NOVBUSSENS_Pos) /*!< 0x00200000 */
+    #define USB_OTG_GCCFG_NOVBUSSENS                 USB_OTG_GCCFG_NOVBUSSENS_Msk  /*!< VBUS sensing disable option */
+
     #define IS_OTG_HS 0
-  #endif
-  #define GPIO_D_NEG GPIO('A', 11)
-  #define GPIO_D_POS GPIO('A', 12)
-  #define GPIO_FUNC GPIO_FUNCTION(10)
-  DECL_CONSTANT_STR("RESERVE_PINS_USB", "PA11,PA12");
+    #define GPIO_D_NEG GPIO('A', 11)
+    #define GPIO_D_POS GPIO('A', 12)
+    #define GPIO_FUNC GPIO_FUNCTION(10)
+    DECL_CONSTANT_STR("RESERVE_PINS_USB", "PA11,PA12");
 #endif
 
 #if IS_OTG_HS
@@ -318,6 +331,23 @@ usb_send_ep0(const void *data, uint_fast8_t len)
     return ret;
 }
 
+// Clear STALL condition on an endpoint
+void
+usb_clear_stall(uint32_t ep, int is_in)
+{
+    usb_irq_disable();
+    if (is_in) {
+        // Clear STALL and set EPENA and CNak
+        EPIN(ep)->DIEPCTL |= (1 << 26) | (1 << 31);  // CNak (bit 26) and EPEna (bit 31)
+        EPIN(ep)->DIEPCTL &= ~(1 << 21);  // Clear STALL bit (bit 21)
+    } else {
+        // Clear STALL and set EPENA and CNak
+        EPOUT(ep)->DOEPCTL |= (1 << 26) | (1 << 31);  // CNak (bit 26) and EPEna (bit 31)
+        EPOUT(ep)->DOEPCTL &= ~(1 << 21);  // Clear STALL bit (bit 21)
+    }
+    usb_irq_enable();
+}
+
 void
 usb_stall_ep0(void)
 {
@@ -410,17 +440,21 @@ OTG_FS_IRQHandler(void)
 void
 usb_init(void)
 {
-    // Enable USB clock
-#if CONFIG_MACH_STM32H7
-    if (READ_BIT(PWR->CR3, PWR_CR3_USB33RDY) != (PWR_CR3_USB33RDY)) {
-        SET_BIT(PWR->CR3, PWR_CR3_USB33DEN);
-    }
-    SET_BIT(RCC->AHB1ENR, USBOTGEN);
+#if !CONFIG_MACH_AT32F415
+        // Enable USB clock
+    #if CONFIG_MACH_STM32H7
+        if (READ_BIT(PWR->CR3, PWR_CR3_USB33RDY) != (PWR_CR3_USB33RDY)) {
+            SET_BIT(PWR->CR3, PWR_CR3_USB33DEN);
+        }
+        SET_BIT(RCC->AHB1ENR, USBOTGEN);
+    #else
+        RCC->AHB2ENR |= RCC_AHB2ENR_OTGFSEN;
+    #endif
+        while (!(OTG->GRSTCTL & USB_OTG_GRSTCTL_AHBIDL))
+        ;
 #else
-    RCC->AHB2ENR |= RCC_AHB2ENR_OTGFSEN;
+    at32f415rc_usbotg_clock_config();
 #endif
-    while (!(OTG->GRSTCTL & USB_OTG_GRSTCTL_AHBIDL))
-        ;
 
     // Configure USB in full-speed device mode
     OTG->GUSBCFG = (USB_OTG_GUSBCFG_FDMOD | USB_OTG_GUSBCFG_PHYSEL
@@ -428,8 +462,10 @@ usb_init(void)
     OTGD->DCFG |= (3 << USB_OTG_DCFG_DSPD_Pos);
 #if CONFIG_MACH_STM32F446 || CONFIG_MACH_STM32H7 || CONFIG_MACH_STM32F7
     OTG->GOTGCTL = USB_OTG_GOTGCTL_BVALOEN | USB_OTG_GOTGCTL_BVALOVAL;
+#elif CONFIG_MACH_AT32F415
+    OTG->GCCFG |= (USB_OTG_GCCFG_VBUSBSEN | USB_OTG_GCCFG_NOVBUSSENS);
 #else
-    OTG->GCCFG |= USB_OTG_GCCFG_NOVBUSSENS;
+    OTG->GCCFG |=  USB_OTG_GCCFG_NOVBUSSENS;
 #endif
 
     // Route pins
